---
title: "Mini-bulk RNA-Seq analysis"
output:
  word_document: default
  pdf_document: default
  html_document: default
---

# Analysis of mini-bulk RNA-Seq samples
From CD4 and CD8 cells, tetramer-sorted against NP16, ORF3a-28, S34 and M24.

## Clustering 

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source('functions.R')
library(org.Hs.eg.db)
library(DESeq2)
library(pheatmap)
library(dplyr)
library(ggplot2)
library(ggrepel)
library (stringi)
library(AnnotationDbi)
library(edgeR)
library(statmod)
library(BiocParallel)
library(gplots)
library (limma)
library("genefilter")
require(data.table)
library(tidyverse)
library(DT)
library(clusterProfiler)
# to convert to Word document, type in Console >rmarkdown::render("file_name.Rmd")
```

```{r}
# Data wrangling to get bulk samples into 1 counts table/expression matrix
setwd('/t1-data/user/lfelce/scRNA-Seq/SmartSeq2_T-cells/')

## load data (from featureCounts)
all_201203 <-  fread('201203_counts.txt', stringsAsFactors = F, header=T)

# remove columns with chromosome, start, end, strand and length info
all_201203 <- all_201203[,-c(2:6)]

# make Geneid into row names
all_201203 <-tibble::column_to_rownames(all_201203, "Geneid")

# tidy up sample names
names(all_201203) <- gsub(x = names(all_201203), pattern = "./", replacement = "")  
names(all_201203) <- gsub(x = names(all_201203), pattern = ".bam", replacement = "")

# list of sample names
list_201203 <- as.data.frame(colnames(all_201203))

## load data (from featureCounts)
all_201221 <-  fread('201221_counts.txt', stringsAsFactors = F, header=T)

# remove columns with chromosome, start, end, strand and length info
all_201221 <- all_201221[,-c(2:6)]

# make Geneid into row names
all_201221 <-tibble::column_to_rownames(all_201221, "Geneid")

# tidy up sample names
names(all_201221) <- gsub(x = names(all_201221), pattern = "./", replacement = "")  
names(all_201221) <- gsub(x = names(all_201221), pattern = ".bam", replacement = "")
names(all_201221) <- gsub(x = names(all_201221), pattern = "1131-TP-1_CD8_NP16_", replacement = "")

# list of sample names
list_201221 <- as.data.frame(colnames(all_201221))

# separate orf3a-28 samples
bulk_orf3a <- all_201203[,c(481:486,774:779, 1404:1408)]

# separate np16, s34 and m24 samples
bulk_np16_s34_m24 <- all_201221[,25:96]

# relabel columns based on patient details
meta_data <- read.table("bulk_metadata.csv", sep=",", header = TRUE)
meta_data_1 <- meta_data[-c(1:17),]
rownames(meta_data_1) <- meta_data_1$Samples
colnames(bulk_np16_s34_m24) <- rownames(meta_data_1)

bulk_np16_s34_m24 <- bulk_np16_s34_m24[,-c(1:3, 6, 13, 18, 55, 56, 67, 68)]

bulk_np16 <- select(bulk_np16_s34_m24,contains("NP16"))
bulk_s34 <- select(bulk_np16_s34_m24,contains("S34"))
bulk_m24 <- select(bulk_np16_s34_m24,contains("M24"))

# separate metadata file
meta_data_2 <- na.omit(meta_data_1)

meta_data_np16 <- meta_data_2 %>% filter(Epitope == "NP16") 
meta_data_orf3a <- meta_data %>% filter(Epitope== "ORF3a-28")
meta_data_s34 <- meta_data_2 %>% filter(Epitope == "S34")
meta_data_m24 <- meta_data_2 %>% filter(Epitope == "M24")

# meta_data <- read.table("bulk_metadata.csv", sep=",", header = TRUE)
# rownames(meta_data) <- meta_data$Samples
# colnames(bulk_data) <- rownames(meta_data)
# all(rownames(meta_data) %in% colnames(bulk_data))
```


```{r, include=FALSE}
dds_np16 <- DESeqDataSetFromMatrix(countData=bulk_np16, colData=meta_data_np16, design=~Tetramer)
keep <- rowSums(counts(dds_np16)) >= 10
dds_np16 <- dds_np16[keep,]
dds_np16 <- DESeq(dds_np16, parallel=TRUE)

dds_orf3a <- DESeqDataSetFromMatrix(countData=bulk_orf3a, colData=meta_data_orf3a, design=~Tetramer)
keep <- rowSums(counts(dds_orf3a)) >= 10
dds_orf3a <- dds_orf3a[keep,]
dds_orf3a <- DESeq(dds_orf3a, parallel=TRUE)

dds_s34 <- DESeqDataSetFromMatrix(countData=bulk_s34, colData=meta_data_s34, design=~Tetramer)
keep <- rowSums(counts(dds_s34)) >= 10
dds_s34 <- dds_s34[keep,]
dds_s34 <- DESeq(dds_s34, parallel=TRUE)

dds_m24 <- DESeqDataSetFromMatrix(countData=bulk_m24, colData=meta_data_m24, design=~Tetramer)
keep <- rowSums(counts(dds_m24)) >= 10
dds_m24 <- dds_m24[keep,]
dds_m24 <- DESeq(dds_m24, parallel=TRUE)
```

### Heatmap of counts matrix

```{r, fig.width=10, fig.height=8}
vsd_np16 <- varianceStabilizingTransformation(dds_np16, blind=FALSE)
select <- order(rowMeans(counts(dds_np16, normalized=FALSE)), decreasing=TRUE)[1:200]
df <- as.data.frame(colData(dds_np16)[,"Tetramer"])
annotation <- data.frame(Var1 = meta_data_np16$Tetramer, Var2 = meta_data_np16$Patient)
rownames(annotation) <- colnames(assay(vsd_np16))
pheatmap(assay(vsd_np16)[select,], cluster_rows = FALSE, show_rownames = FALSE, cluster_cols = TRUE, annotation =annotation)

vsd_orf3a <- varianceStabilizingTransformation(dds_orf3a, blind=FALSE)
select <- order(rowMeans(counts(dds_orf3a, normalized=FALSE)), decreasing=TRUE)[1:200]
df <- as.data.frame(colData(dds_orf3a)[,"Tetramer"])
annotation <- data.frame(Var1 = meta_data_orf3a$Tetramer, Var2 = meta_data_orf3a$Patient)
rownames(annotation) <- colnames(assay(vsd_orf3a))
pheatmap(assay(vsd_orf3a)[select,], cluster_rows = FALSE, show_rownames = FALSE, cluster_cols = TRUE, annotation =annotation)

vsd_s34 <- varianceStabilizingTransformation(dds_s34, blind=FALSE)
select <- order(rowMeans(counts(dds_s34, normalized=FALSE)), decreasing=TRUE)[1:200]
df <- as.data.frame(colData(dds_s34)[,"Tetramer"])
annotation <- data.frame(Var1 = meta_data_s34$Tetramer, Var2 = meta_data_s34$Patient)
rownames(annotation) <- colnames(assay(vsd_s34))
pheatmap(assay(vsd_s34)[select,], cluster_rows = FALSE, show_rownames = FALSE, cluster_cols = TRUE, annotation =annotation)

vsd_m24 <- varianceStabilizingTransformation(dds_m24, blind=FALSE)
select <- order(rowMeans(counts(dds_m24, normalized=FALSE)), decreasing=TRUE)[1:200]
df <- as.data.frame(colData(dds_m24)[,"Tetramer"])
annotation <- data.frame(Var1 = meta_data_m24$Tetramer, Var2 = meta_data_m24$Patient)
rownames(annotation) <- colnames(assay(vsd_m24))
pheatmap(assay(vsd_m24)[select,], cluster_rows = FALSE, show_rownames = FALSE, cluster_cols = TRUE, annotation =annotation)


```



### PCA analysis of the samples 

#### PCA - NP16

```{r, fig.width=6, fig.height=4}
plotPCA(vsd_np16, intgroup=c("Patient"))
plotPCA(vsd_np16, intgroup=c("Tetramer"))
```
#### PCA - ORF3a-28

```{r, fig.width=8, fig.height=3}
plotPCA(vsd_orf3a, intgroup=c("Patient"))
plotPCA(vsd_orf3a, intgroup=c("Tetramer"))
```

#### PCA - S34

```{r, fig.width=6, fig.height=3}
plotPCA(vsd_s34, intgroup=c("Patient"))
plotPCA(vsd_s34, intgroup=c("Tetramer"))
```
#### PCA - M24

```{r, fig.width=6, fig.height=3}
plotPCA(vsd_m24, intgroup=c("Patient"))
plotPCA(vsd_m24, intgroup=c("Tetramer"))
```


### Heatmap of gene clustering

```{r, fig.width=10, fig.height=8}
topVarGenes <- head(order(rowVars(assay(vsd_np16)), decreasing = TRUE), 20)
mat  <- assay(vsd_np16)[ topVarGenes, ]
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vsd_np16)[, c("Patient","Tetramer")])
pheatmap(mat, annotation_col = anno)

topVarGenes <- head(order(rowVars(assay(vsd_orf3a)), decreasing = TRUE), 20)
mat  <- assay(vsd_orf3a)[ topVarGenes, ]
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vsd_orf3a)[, c("Patient","Tetramer")])
pheatmap(mat, annotation_col = anno)

topVarGenes <- head(order(rowVars(assay(vsd_s34)), decreasing = TRUE), 20)
mat  <- assay(vsd_s34)[ topVarGenes, ]
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vsd_s34)[, c("Patient","Tetramer")])
pheatmap(mat, annotation_col = anno)

topVarGenes <- head(order(rowVars(assay(vsd_m24)), decreasing = TRUE), 20)
mat  <- assay(vsd_m24)[ topVarGenes, ]
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vsd_m24)[, c("Patient","Tetramer")])
pheatmap(mat, annotation_col = anno)

```

## Differential Gene Expression Analysis (DESeq2)

```{r}
# set up meta_data for all experiments (only have 1 comparison/metadata sheet)
meta_data_np16 <- meta_data %>% filter(Epitope == "NP16")
meta_data_np16 <- na.omit(meta_data_np16)
meta_data_s34 <- meta_data %>% filter(Epitope == "S34")
meta_data_s34 <- na.omit(meta_data_s34)
meta_data_m24 <- meta_data %>% filter(Epitope == "M24")
meta_data_m24 <- na.omit(meta_data_m24)
```



```{r, message=FALSE, eval=FALSE}
# Set up the DESeq2 object
df_mRNA <-  bulk_data
run_deseq2 <- function(df_mRNA, meta_data, control="Negative", test="Positive", value){
  
  df_mRNA = df_mRNA[,rownames(meta_data)]
  
  
  dds<- DESeqDataSetFromMatrix(countData=df_mRNA,
                               colData=meta_data,
                               design= ~Tetramer)
  
  keep <- rowSums(counts(dds)) >= 10
  dds <- dds[keep,]
  
  dds <- DESeq(dds)
  
  res <- results(dds, contrast = c(value, test,control))
  
  return(res)
}
res_np16 <- run_deseq2(df_mRNA, meta_data_np16, control = "Negative", test="Positive", value="Tetramer")
res_s34 <- run_deseq2(df_mRNA, meta_data_s34, control = "Negative", test="Positive", value="Tetramer")
res_m24 <- run_deseq2(df_mRNA, meta_data_m24, control = "Negative", test="Positive", value="Tetramer")
```

The following describes the analysis of the data using DEseq2. I have separated the analyses into three different contrasts and have plotted summary statistics and MA plots for each.  


### Summary of the data

#### CD8+ NP16-specific T cells, negative vs positive tetramer sorting
Each sample from each patient is used as biological replicate

```{r}
summary(res_np16, alpha=0.05)
```
#### CD4+ S34-specific T cells, negative vs positive tetramer sorting
Each sample from each patient is used as biological replicate

```{r}
summary(res_s34, alpha=0.05)
```
#### CD4+ M24-specific T cells, negative vs positive tetramer sorting
Each sample from each patient is used as biological replicate

```{r}
summary(res_m24, alpha=0.05)
```
### MA plots 

In DESeq2, the function plotMA shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet. Points will be colored blue if the adjusted p value is less than 0.01. Points which fall out of the window are plotted as open triangles pointing either up or down.

#### CD8+ NP16-specific T cells, negative vs positive tetramer sorting

```{r, fig.width=6, fig.height=4}
DESeq2::plotMA(res_np16)
```
#### CD4+ S34-specific T cells, negative vs positive tetramer sorting

```{r, fig.width=6, fig.height=4}
DESeq2::plotMA(res_s34)
```
#### CD4+ M24-specific T cells, negative vs positive tetramer sorting

```{r, fig.width=6, fig.height=4}
DESeq2::plotMA(res_m24)
```

### Volcano plots 

#### CD8+ NP16-specific T cells, negative vs positive tetramer sorting

```{r, message=FALSE, , fig.width=6, fig.height=6}
res <- as.data.frame(res_np16)
mutateddf <- mutate(res, sig=ifelse(res$padj<0.01, "P.adj<0.01", "Not Sig")) #Will have different colors depending on significance
input <- cbind(gene=rownames(res), mutateddf )
input <- input %>%
  arrange(input$padj)
#convert the rownames to a column
volc = ggplot(input, aes(log2FoldChange, -log10(padj))) + #volcanoplot with log2Foldchange versus pvalue
    geom_point(aes(col=sig)) + #add points colored by significance
    ggtitle("CD8+ NP16-tetramer positive T cells") #e.g. 'Volcanoplot DESeq2'
volc
```
#### CD4+ S34-specific T cells, negative vs positive tetramer sorting

```{r, message=FALSE, , fig.width=6, fig.height=6}
res <- as.data.frame(res_s34)
mutateddf <- mutate(res, sig=ifelse(res$padj<0.01, "P.adj<0.01", "Not Sig")) #Will have different colors depending on significance
input <- cbind(gene=rownames(res), mutateddf )
input <- input %>%
  arrange(input$padj)
#convert the rownames to a column
volc = ggplot(input, aes(log2FoldChange, -log10(padj))) + #volcanoplot with log2Foldchange versus pvalue
    geom_point(aes(col=sig)) + #add points colored by significance
    ggtitle("CD8+ NP16-tetramer positive T cells") #e.g. 'Volcanoplot DESeq2'
volc
```
#### CD4+ M24-specific T cells, negative vs positive tetramer sorting

```{r, message=FALSE, , fig.width=6, fig.height=6}
res <- as.data.frame(res_m24)
mutateddf <- mutate(res, sig=ifelse(res$padj<0.01, "P.adj<0.01", "Not Sig")) #Will have different colors depending on significance
input <- cbind(gene=rownames(res), mutateddf )
input <- input %>%
  arrange(input$padj)
#convert the rownames to a column
volc = ggplot(input, aes(log2FoldChange, -log10(padj))) + #volcanoplot with log2Foldchange versus pvalue
    geom_point(aes(col=sig)) + #add points colored by significance
    ggtitle("CD8+ NP16-tetramer positive T cells") #e.g. 'Volcanoplot DESeq2'
volc
```


### Results tables

The folowing results tables show the significant genes (p<0.05). Filtering has been performed with a log2 fold change +/- 2. These have been exported as .csv files where _res.csv is all genes, and _sig.csv is significant genes (p<0.05) with LFC +/-2 (see results folder)


#### CD8+ NP16-specific T cells, negative vs positive tetramer sorting

```{r, message=FALSE}
# modify filter_genes function as already have gene names set as row names rather than Ensembl IDs which have to be converted to gene symbols.
dt <- filter_genes(as.data.frame(res_np16), name="cd8_np16")
datatable(dt$sig)
# quick check using Calli's marker list
np16_markers <- dt$sig[is.element(dt$sig$Gene,tmarkers$V1),]
```

#### CD4+ S34-specific T cells, negative vs positive tetramer sorting

```{r, message=FALSE}
# modify filter_genes function as already have gene names set as row names rather than Ensembl IDs which have to be converted to gene symbols.
dt <- filter_genes(as.data.frame(res_s34), name="cd4_s34")
datatable(dt$sig)
# quick check using Calli's marker list
s34_markers <- dt$sig[is.element(dt$sig$Gene,tmarkers$V1),]
```

#### CD4+ M24-specific T cells, negative vs positive tetramer sorting

```{r, message=FALSE}
# modify filter_genes function as already have gene names set as row names rather than Ensembl IDs which have to be converted to gene symbols.
dt <- filter_genes(as.data.frame(res_m24), name="cd4_m24")
datatable(dt$sig)
# quick check using Calli's marker list
m24_markers <- dt$sig[is.element(dt$sig$Gene,tmarkers$V1),]
```

## GO Analysis

#### CD8+ NP16-specific T cells, negative vs positive tetramer sorting
```{r}
# if just use significant DEGs - only 155, not enough for GO analysis.
# use all DEGs
deg_np16 <- read.csv('/t1-data/user/lfelce/scRNA-Seq/SmartSeq2_T-cells/results/cd8_np16_res.csv')
deg_np16 <- deg_np16 %>% filter (padj < 0.5)
deg_np16$Entrez.Gene <- mapIds(org.Hs.eg.db, keys=as.character(deg_np16$Gene), keytype="SYMBOL", column="ENTREZID")
# Keep only rows from table without NAs
deg_np16 <- deg_np16[is.na(deg_np16$Entrez.Gene)==FALSE,]
# Remove duplicated entries
deg_np16 <- deg_np16[!duplicated(deg_np16$Entrez.Gene),]
# Change Entrez IDs from numbers to characters
geneset <- as.character(deg_np16$Entrez.Gene)
# This will take a little while to run
ego <- enrichGO(gene = geneset, 
                universe = NULL, #all available genes in database
                OrgDb = org.Hs.eg.db, #Hs: homo sapiens
                ont ="BP", #molecular function, biological process, cellular component
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,  # q value is FDR adjusted p value
                readable = TRUE) #will show gene symbol in images later rather than Entrez Gene ID
# dimensions - number of GO terms
dim(ego)
simp <- simplify(ego)
np16_GOgenes <- data.frame(simp$ID, simp$Description, simp$p.adjust, simp$geneID)
# Copy and paste table into Excel
```

```{r,fig.height=4, fig.width=6}
barplot (simp, showCategory = 10)
```

```{r, fig.height=8, fig.width=10}
emapplot(simp)
```

```{r, fig.height=12, fig.width=14}
# logFC is optional, but if available will colour-code the nodes
logFC_np16 <- deg_np16$log2FoldChange
# set name of object
names(logFC_np16) <- deg_np16$Entrez.Gene
 
cnetplot(simp, foldChange = logFC_np16, 
         showCategory = c("T cell activation",
                          "response to virus", 
                          "leukocyte cell-cell adhesion", 
                          "T cell receptor signaling pathway", 
                          "cell killing", 
                          "interferon-gamma production",
                         "lymphocyte proliferation"), 
         circular = FALSE)
# show 5 GO terms, if set circular to TRUE then GO terms won't be in middle, will be on edge of circle
# size = number of genes, fold change = red for upregulated, green for downregulated
```
#### CD4+ S34-specific T cells, negative vs positive tetramer sorting
```{r}
deg_s34 <- read.csv('/t1-data/user/lfelce/scRNA-Seq/SmartSeq2_T-cells/results/cd4_s34_res.csv')
deg_s34$Entrez.Gene <- mapIds(org.Hs.eg.db, keys=as.character(deg_s34$Gene), keytype="SYMBOL", column="ENTREZID")
# Keep only rows from table without NAs
deg_s34 <- deg_s34[is.na(deg_s34$Entrez.Gene)==FALSE,]
# Remove duplicated entries
deg_s34 <- deg_s34[!duplicated(deg_s34$Entrez.Gene),]
# Change Entrez IDs from numbers to characters
geneset <- as.character(deg_s34$Entrez.Gene)
# This will take a little while to run
ego <- enrichGO(gene = geneset, 
                universe = NULL, #all available genes in database
                OrgDb = org.Hs.eg.db, #Hs: homo sapiens
                ont ="BP", #molecular function, biological process, cellular component
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,  # q value is FDR adjusted p value
                readable = TRUE) #will show gene symbol in images later rather than Entrez Gene ID
# dimensions - number of GO terms
dim(ego)
simp <- simplify(ego)
dim(simp)
s34_GOgenes <- data.frame(simp$ID, simp$Description, simp$p.adjust, simp$geneID)
# Copy and paste table into Excel
```

```{r, fig.height=12, fig.width=14}
# logFC is optional, but if available will colour-code the nodes
logFC_s34 <- deg_s34$log2FoldChange
# set name of object
names(logFC_s34) <- deg_s34$Entrez.Gene
 
cnetplot(simp, foldChange = logFC_s34, 
         showCategory = c("T cell activation",
                         "T cell receptor signaling pathway",
                         "leukocyte cell-cell adhesion",
                         "type I interferon production",
                         "B cell activation"), 
         circular = FALSE)
# show 5 GO terms, if set circular to TRUE then GO terms won't be in middle, will be on edge of circle
# size = number of genes, fold change = red for upregulated, green for downregulated
```


#### CD4+ M24-specific T cells, negative vs positive tetramer sorting
```{r}
deg_m24 <- read.csv('/t1-data/user/lfelce/scRNA-Seq/SmartSeq2_T-cells/results/cd4_m24_sig.csv')
deg_m24 <- deg_m24 %>% filter(log2FoldChange < -3 | log2FoldChange > 3)
deg_m24$Entrez.Gene <- mapIds(org.Hs.eg.db, keys=as.character(deg_m24$Gene), keytype="SYMBOL", column="ENTREZID")
# Keep only rows from table without NAs
deg_m24 <- deg_m24[is.na(deg_m24$Entrez.Gene)==FALSE,]
# Remove duplicated entries
deg_m24 <- deg_m24[!duplicated(deg_m24$Entrez.Gene),]
# Change Entrez IDs from numbers to characters
geneset <- as.character(deg_m24$Entrez.Gene)
# This will take a little while to run
ego <- enrichGO(gene = geneset, 
                universe = NULL, # all available genes in database
                OrgDb = org.Hs.eg.db, # Hs: homo sapiens
                ont ="BP", # molecular function, biological process, cellular component
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,  # q value is FDR adjusted p value
                readable = TRUE) # will show gene symbol in images later rather than Entrez Gene ID
# dimensions - number of GO terms
dim(ego)
simp <- simplify(ego)
dim(simp)
m24_GOgenes <- data.frame(simp$ID, simp$Description, simp$p.adjust, simp$geneID)
# Copy and paste table into Excel
```

```{r,fig.height=4, fig.width=6}
barplot (simp, showCategory = 10)
```
```{r, fig.height=8, fig.width=10}
emapplot(simp, showCategory = 30)
```

```{r, fig.height=12, fig.width=14}
# logFC is optional, but if available will colour-code the nodes
logFC_m24 <- deg_m24$log2FoldChange
# set name of object
names(logFC_m24) <- deg_m24$Entrez.Gene
 
cnetplot(simp, foldChange = logFC_m24, 
         showCategory = 5, 
         circular = FALSE)
# show 5 GO terms, if set circular to TRUE then GO terms won't be in middle, will be on edge of circle
# size = number of genes, fold change = red for upregulated, green for downregulated
```
