---
title: "SARS-CoV-2 6 month T cells"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(SingleCellExperiment)
library(scater)
library(DropletUtils)
library(scran)


```

## Analysis of 10x 6 month convalescent Dong050121 tube 2 samples using Single Cell Experiment
Following this guide http://bioconductor.org/books/release/OSCA/droplet-processing.html#droplet-processing but leaving out removeAmbience as not provided in older version of DropletUtils currently installed on cluster

### Read in 10X Genomics data from UMI count matrix
```{r}
# read in data and create Single Cell Experiment object
# have to use raw unfiltered matrix if going to do cell calling from scratch
# without using CellRanger's cell calling
t2 <- read10xCounts("/t1-data/user/lfelce/10x_Dong050121/combined_T2_repeat4/outs/raw_feature_bc_matrix", type="sparse")

# set Antibody Capture (cell hashing) as separate alternative experiment within SingleCellExperiment object
t2.type <- as.data.frame(rowData(t2))

is.hto <- grepl("Antibody Capture", t2.type$Type)

test <- splitAltExps(t2, ifelse(is.hto, "Antibody Capture", "Gene Expression"))

altExpNames(test)

sce <- test

```

### Computing barcode ranks
Examine distribution of total counts across barcodes, focusing on those with largest counts
```{r, fig.height=4, fig.width=5}
bcrank <- barcodeRanks(counts(sce))

# Only showing unique points for plotting speed.
uniq <- !duplicated(bcrank$rank)
plot(bcrank$rank[uniq], bcrank$total[uniq], log="xy",
    xlab="Rank", ylab="Total UMI count", cex.lab=1.2)

abline(h=metadata(bcrank)$inflection, col="darkgreen", lty=2)
abline(h=metadata(bcrank)$knee, col="dodgerblue", lty=2)

legend("bottomleft", legend=c("Inflection", "Knee"), 
        col=c("darkgreen", "dodgerblue"), lty=2, cex=1.2)

```
Knee and inflection points mark transition between two components of total count distribution. Represents difference between empty droplets with little RNA and cell containing droplets with more RNA.

### Detecting empty droplets
Empty droplets contain RNA from ambient solution, resulting in non-zero counts after debarcoding. Use emptyDrops to distinguish between emptry droplets and cells by testing each barcode's expression profile for significant deviation from ambient profile
```{r, fig.height=4, fig.width=6}
# emptyDrops performs Monte Carlo simulations to compute p-values,
# so we need to set the seed to obtain reproducible results.
set.seed(100)
e.out <- emptyDrops(counts(sce))

# See ?emptyDrops for an explanation of why there are NA values.
summary(e.out$FDR <= 0.001)

# significant true and non-limited true - number of cells not including empty droplets?
table(Sig=e.out$FDR <= 0.001, Limited=e.out$Limited)

# check distribution of p-values for low-total barcodes - distribution should be close to uniform.
# large peaks near 0 indicate that barcodes with total counts below value set at limit are not all ambient in origin
# decrease limit to ensure barcodes corresponding to droplets with very small cells are not used to estimate ambient profile
set.seed(100)
limit <- 20   
all.out <- emptyDrops(counts(sce), lower=limit, test.ambient=TRUE)
hist(all.out$PValue[all.out$Total <= limit & all.out$Total > 0],
    xlab="P-value", main="", col="grey80")

# once happy with performance of emptyDrops, subset SingleCellExperiment object to retain only detected cells
sce <- sce[,which(e.out$FDR <= 0.001)]
```

### Demultiplexing cell hashes
