---
title: "SARS-CoV-2 6 month T cells"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(plyr)
library(dplyr)
library(patchwork)
library(cowplot)
library(ggplot2)
library(data.table)
library(tidyverse)
library(harmony)
library(gridExtra)
```

# 10x single cell RNA sequencing from SARS-CoV-2 CD4/8 tetramer sorted T cells

## Batch 1
Dong 171220 = batch1
Hashtag Ab 1 1180_6Mon, expanded, DR0101_M24 tetramer, CD4
Hashtag Ab 2 1344_6Mon, expanded, DR1501_S51 tetramer, CD4
Hashtag Ab 4 1344_6Mon, expanded, DR1501_S174 tetramer, CD4
Hashtag Ab 5 1180_6Mon, expanded, A3_S76 tetramer, CD8
Hashtag Ab 6 1180_6Mon, expanded, A1-ORF3a-28 tetramer, CD8
Hashtag Ab 7 1344_6Mon, expanded, B27_NP1 tetramer, CD8
Hashtag Ab 8 1344_6Mon, expanded, B27_NP13 tetramer, CD8
Note: hashtag 5 and 6 mixed up?

### Separate Gene Expression and Antibody Capture
```{r}
# # Read in data
# batch1.counts <- Read10X(data.dir = "/t1-data/user/lfelce/10x_DONG171220/counts_gex/filtered_feature_bc_matrix")
# colnames(batch1.counts) <- gsub(x = colnames(batch1.counts), pattern = "-1", replacement = "")
# 
# batch1.hash <- Read10X(data.dir ="/t1-data/user/lfelce/10x_DONG171220/counts_ab/filtered_feature_bc_matrix")
# colnames(batch1.hash) <- gsub(x = colnames(batch1.hash), pattern = "-1", replacement = "")
# 
# # Select cell barcodes detected by both RNA and hashtag, then filter
# joint.barcodes <- intersect(colnames(batch1.counts), colnames(batch1.hash))
# 
# # Subset RNA and hashtag counts by joint cell barcodes
# batch1.counts <- batch1.counts[, joint.barcodes]
# batch1.hash <- as.matrix(batch1.hash[,joint.barcodes])
# 
# # remove Hashtag3 (no cells labelled with this Ab)
# batch1.hash <- batch1.hash[-3,]
# 
# # rename rows to sample names
# rownames(batch1.hash) <- c("1180_CD4_M24", "1344_CD4_S51", "1344_CD4_S174", "1180_CD8_S76", "1180_CD8_ORF3a-28", "1344_CD8_NP1", "1344_CD8_NP13")
# 
# # counts per hashtag
# rowSums(batch1.hash)
# rowMeans(batch1.hash)
# 
# # Set up Seurat object
# batch1 <- CreateSeuratObject(counts = batch1.counts)
# 
# # Normalize RNA data with log normalization
# batch1 <- NormalizeData(batch1)
# 
# # Find and scale variable features
# batch1 <- FindVariableFeatures(batch1, selection.method = "mean.var.plot")
# 
# batch1 <- ScaleData(batch1, features = VariableFeatures(batch1))
# 
# # Add HTO data as a new assay independent from RNA
# batch1[["hashtag"]] <- CreateAssayObject(counts = batch1.hash)
# 
# # Normalize HTO data, here we use centered log-ratio (CLR) transformation
# batch1 <- NormalizeData(batch1, assay = "hashtag", normalization.method = "CLR")
# 
# # Demultiplex cells based on hashtag enrichment - use HTODemux() to assign single cells back to sample origins
# 
# # If you have a very large dataset we suggest using k_function = 'clara'. This is a k-medoid
# # clustering function for large applications You can also play with additional parameters (see
# # documentation for HTODemux()) to adjust the threshold for classification Here we are using the
# # default settings
# batch1 <- HTODemux(batch1, assay = "hashtag", positive.quantile = 0.99)

```

Output from running HTODemux is saved in metadata. Can visualise cells classified as singlets, doublets and negative/ambiguous cells

```{r, fig.width=12, fig.height=12}
# # Global classification results
# table(batch1$hashtag_classification.global)
# 
# # Visualise enrichment for selected hashtags with ridge plots
# # Group cells based on max hashtag signal
# Idents(batch1) <- "hashtag_maxID"
# RidgePlot(batch1, assay="hashtag", features=rownames(batch1[["hashtag"]]), ncol=2)

```

```{r, fig.width=6, fig.height=4}
# # Compare number of UMIs for singlets, doublets and negative cells
# Idents(batch1) <- "hashtag_classification.global"
# VlnPlot(batch1, features = "nCount_RNA", pt.size = 0.1, log = TRUE)
```
#### Singlets only
```{r, fig.width=20, fig.height=24}
# # Cluster and visualize cells using the usual scRNA-seq workflow, and examine for the potential presence of batch effects.
# 
# Extract the singlets
# batch1.singlet <- subset(batch1, idents = "Singlet")

# # Select the top 1000 most variable features
# batch1.singlet <- FindVariableFeatures(batch1.singlet, selection.method = "mean.var.plot")
# 
# # Scaling RNA data
# batch1.singlet <- ScaleData(batch1.singlet, features = VariableFeatures(batch1.singlet))
# 
# # Run PCA
# batch1.singlet <- RunPCA(batch1.singlet, features = VariableFeatures(batch1.singlet))
# 
# # Select top PCs for clustering
# batch1.singlet <- FindNeighbors(batch1.singlet, reduction = "pca", dims = 1:20)
# batch1.singlet <- FindClusters(batch1.singlet, resolution = 0.5, verbose = FALSE)
# batch1.singlet <- RunUMAP(batch1.singlet, dims = 1:20)
# batch1.singlet <- RunTSNE(batch1.singlet, reduction = "pca", dims = 1:20)
# 
# # Assign metadata to object
# patient_list <- as.data.frame(batch1.singlet@meta.data$hashtag_classification)
# colnames(patient_list) <- "sample"
# patient_list$patient <- ifelse(grepl("1180", patient_list$sample), "1180",
#                                ifelse(grepl("1344", patient_list$sample), "1344", ""))
# patient_list$subset <- ifelse(grepl("CD4", patient_list$sample), "CD4",
#                                ifelse(grepl("CD8", patient_list$sample), "CD8", ""))
# patient_list$epitope <- ifelse(grepl("M24", patient_list$sample), "membrane",
#                         ifelse(grepl("S51", patient_list$sample), "spike",
#                         ifelse(grepl("S174", patient_list$sample), "spike",
#                         ifelse(grepl("S76", patient_list$sample), "spike",
#                         ifelse(grepl("ORF3a-28", patient_list$sample), "ORF3a",
#                         ifelse(grepl("NP1", patient_list$sample), "nucleoprotein",
#                         ifelse(grepl("NP13", patient_list$sample), "nucleoprotein", "")))))))
# 
# batch1.singlet@meta.data$"epitope" <- as.factor(patient_list$epitope)
# batch1.singlet@meta.data$"patient" <- as.factor(patient_list$patient)
# batch1.singlet@meta.data$"t-cell" <- as.factor(patient_list$subset)
# # 
# # Projecting singlet identities on UMAP visualization
# tsne <- DimPlot(batch1.singlet, reduction="tsne", group.by = "hashtag_classification")
# pca <- DimPlot(batch1.singlet, reduction="pca", group.by="hashtag_classification")
# pca2 <- DimPlot(batch1.singlet, reduction="pca", group.by="epitope")
# 
# umap_cluster <- DimPlot(batch1.singlet, reduction="umap", group.by = "seurat_clusters", label=TRUE)
# umap_tcell <- DimPlot(batch1.singlet, reduction="umap", group.by = "t-cell")
# umap_epitope <- DimPlot(batch1.singlet, reduction="umap", group.by = "epitope")
# umap_patient <- DimPlot(batch1.singlet, reduction="umap", group.by = "patient")
# umap_hash <- DimPlot(batch1.singlet, reduction="umap", group.by="hashtag_classification")
# 
# grid.arrange(tsne, pca, pca2, umap_cluster, umap_tcell, umap_epitope, umap_patient, umap_hash, nrow=4)
# 
# 
# # Save batch1 object
# saveRDS(batch1.singlet, "batch1_seurat.rds")
```
#### Cluster markers
```{r}
# batch1.markers <- FindAllMarkers(object = batch1, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
# cluster_0 <- batch1.markers[batch1.markers$cluster == "0", ]
# cluster_1 <- batch1.markers[batch1.markers$cluster == "1", ]
# cluster_2 <- batch1.markers[batch1.markers$cluster == "2", ]
# cluster_3 <- batch1.markers[batch1.markers$cluster == "3", ]
# cluster_4 <- batch1.markers[batch1.markers$cluster == "4", ]
# cluster_5 <- batch1.markers[batch1.markers$cluster == "5", ]
# cluster_6 <- batch1.markers[batch1.markers$cluster == "6", ]
# cluster_7 <- batch1.markers[batch1.markers$cluster == "7", ]
# cluster_8 <- batch1.markers[batch1.markers$cluster == "8", ]
# 
# setwd('/t1-data/user/lfelce/scRNA-Seq/SmartSeq2_T-cells/')
# tmarkers <- read.csv("Markers_for_Ling.csv", header=F)
# 
# genes_0 <- tmarkers[is.element(tmarkers$V1, cluster_0$gene),]
# genes_1 <- tmarkers[is.element(tmarkers$V1, cluster_1$gene),]
# genes_2 <- tmarkers[is.element(tmarkers$V1, cluster_2$gene),]
# genes_3 <- tmarkers[is.element(tmarkers$V1, cluster_3$gene),]
# genes_4 <- tmarkers[is.element(tmarkers$V1, cluster_4$gene),]
# genes_5 <- tmarkers[is.element(tmarkers$V1, cluster_5$gene),]
# genes_6 <- tmarkers[is.element(tmarkers$V1, cluster_6$gene),]
# genes_7 <- tmarkers[is.element(tmarkers$V1, cluster_7$gene),]
# genes_8 <- tmarkers[is.element(tmarkers$V1, cluster_8$gene),]
# 
# genes <- read.csv("genes.csv", header=TRUE)

# find cluster marks distinguishing cluster 0 from clusters 1 and 7 (spike only clusters)
# cluster0.markers <- FindMarkers(batch1, ident.1 = 0, ident.2 = c(1,7), min.pct = 0.25)

# cluster0.markers <- FindMarkers(batch1, ident.1 = 0, ident.2 = c(1, 2, 3, 4, 5, 6, 7, 8), min.pct = 0.25)

```
#### Gene expression by T cell subset
```{r, fig.width=20, fig.height=36}
# batch1 <- readRDS("batch1_seurat.rds")
# 
# p1 <- VlnPlot(object = batch1, features=c("IFNG", "TNF", "TGFB1", "PTPRC", 
#                                           "CD44", "IL7R", "S100A4", "CD38", 
#                                           "CD27", "CD28", "CD69", "HLA-DRB1", 
#                                           "ITGAE", "ITGB1", "ITGA4", "ICAM3", 
#                                           "ITGB7", "SELPLG","ITGAL","ITGB2",
#                                           "CCL3", "CCL4", "CCL5", "CCR4", 
#                                           "CCR5", "CXCR3", "CXCR4", "CXCR6",
#                                           "MX1", "ISG15", "GZMA", "GZMB", 
#                                           "GZMH", "GZMK", "GZMM", "PRF1", 
#                                           "CSF2", "GNLY", "PDCD1", "KLRG1"),
#         ncol=4, group.by = "t-cell")
# 
# p1
```

#### Gene expression by epitope
```{r, fig.width=20, fig.height=40}
# p2 <- VlnPlot(object = batch1, features=c("IFNG", "TNF", "TGFB1", "PTPRC", 
#                                           "CD44", "IL7R", "S100A4", "CD38", 
#                                           "CD27", "CD28", "CD69", "HLA-DRB1",
#                                           "ITGAE", "ITGB1", "ITGA4", "ICAM3", 
#                                           "ITGB7", "SELPLG", "ITGAL", "ITGB2",
#                                           "CCL3", "CCL4", "CCL5", "CCR4", 
#                                           "CCR5", "CXCR3", "CXCR4", "CXCR6",
#                                           "ISG15", "GZMA", "GZMB", "GZMH", 
#                                           "GZMK", "GZMM", "PRF1", "GNLY",
#                                           "CSF2", "HAVCR2", "KLRG1", "TIGIT"),
#         ncol=4, group.by = "epitope")
# 
# p2
```


#### Gene expression by cluster

```{r, fig.width=30, fig.height=40}
# p3 <- VlnPlot(object = batch1, features=c("IFNG", "TNF", "TGFB1", "PTPRC", 
#                                           "CD44", "IL7R", "S100A4", "CD38", 
#                                           "CD27", "CD28", "CD69", "HLA-DRB1",
#                                           "ITGAE", "ITGB1", "ITGA4", "ICAM3", 
#                                           "ITGB7", "SELPLG", "ITGAL", "ITGB2",
#                                           "CCL3", "CCL4", "CCL5", "CCR4", 
#                                           "CCR5", "CXCR3", "CXCR4", "CXCR6",
#                                           "ISG15", "GZMA", "GZMB", "GZMH", 
#                                           "GZMK", "GZMM", "PRF1", "GNLY",
#                                           "CSF2", "HAVCR2", "KLRG1", "TIGIT"),
#         ncol=4, group.by = "seurat_clusters")
# 
# p3
```

```{r, fig.height=40, fig.width=30}
# p4 <- FeaturePlot(object = batch1, features = c("IFNG", "TNF", "TGFB1", "PTPRC", 
#                                           "CD44", "IL7R", "S100A4", "CD38", 
#                                           "CD27", "CD28", "CD69", "HLA-DRB1",
#                                           "ITGAE", "ITGB1", "ITGA4", "ICAM3", 
#                                           "ITGB7", "SELPLG", "ITGAL", "ITGB2",
#                                           "CCL3", "CCL4", "CCL5", "CCR4", 
#                                           "CCR5", "CXCR3", "CXCR4", "CXCR6",
#                                           "ISG15", "GZMA", "GZMB", "GZMH", 
#                                           "GZMK", "GZMM", "PRF1", "GNLY",
#                                           "CSF2", "HAVCR2", "KLRG1", "TIGIT"), 
#             cols = c("grey", "blue"), reduction = "umap")
# 
# p4
```

```{r, fig.width=16, fig.height=20}
# batch1 <- readRDS("batch1_seurat.rds")
# 
# VlnPlot(object = batch1, features=c("RPS8", "RPLP1", "RPL10", "RPL8",
#                                     "GZMA", "GZMB", "FASLG", "PTPRC", 
#                                     "IFNG", "TNF", "CD8A", "CD8B",
#                                     "CCL3", "CCL4", "CCL5", "CCL3L1", 
#                                     "HLA-A", "HLA-B", "HLA-C", "HLA-F",
#                                     "HLA-DQA1", "HLA-DPB1", "HLA-DRB1", "IL10RA",
#                                     "ITGB2", "ITGA4", "ITGAL", "CD2"),
#         ncol=4, group.by = "seurat_clusters")


```

### Integrated with flu

```{r}
# covid <- batch1.singlet
# 
# setwd('/t1-data/user/lfelce/scRNA-Seq/SmartSeq2_Flu/')
# # load data (from featureCounts)
# d1_counts <-  fread('D1_210121_counts.txt', stringsAsFactors = F, header=T)
# d2_counts <-  fread('D2_210121_counts.txt', stringsAsFactors = F, header=T)
# 
# # remove columns with chromosome, start, end, strand and length info
# d1_counts <- d1_counts[,-c(2:6)]
# d2_counts <- d2_counts[,-c(2:6)]
# 
# # make Geneid into row names
# d1_counts <-tibble::column_to_rownames(d1_counts, "Geneid")
# d2_counts <-tibble::column_to_rownames(d2_counts, "Geneid")
# 
# # tidy up sample names
# names(d1_counts) <- gsub(x = names(d1_counts), pattern = "./", replacement = "")
# names(d1_counts) <- gsub(x = names(d1_counts), pattern = ".bam", replacement = "")
# names(d2_counts) <- gsub(x = names(d2_counts), pattern = "./", replacement = "")
# names(d2_counts) <- gsub(x = names(d2_counts), pattern = ".bam", replacement = "")
# 
# # rename file names based on sample names
# d1_names <- read.csv("donor1_names.csv", header=F)
# d1_names$V2 <- gsub(x = d1_names$V2, pattern = "Sample:", replacement = "")
# d1_names$V2 <- paste("D1", d1_names$V2, sep="_")
# colnames(d1_counts) <- d1_names$V2
# 
# d2_names <- read.csv("donor2_names.csv", header=F)
# d2_names$V2 <- paste("D2", d2_names$V2, sep="_")
# colnames(d2_counts) <- d2_names$V2
# 
# # combine columns
# flu_counts <- cbind(d1_counts, d2_counts)
# 
# # # Initialise Seurat object with raw (non-normalised?) data
# # # min.cells = include features detected in at least this many cells
# # # min.features includes cells where at least this many features detected
# 
# flu <- CreateSeuratObject(counts = flu_counts, min.cells = 3, min.features = 200, project = "Flu CD8+ T-cell data", assay = "RNA")
# 
# combo <- merge(covid, flu)
# 
# covid_list <- as.data.frame(covid@meta.data$hashtag_classification)
# colnames(covid_list) <- "sample"
# 
# covid_list$patient <- ifelse(grepl("1180", covid_list$sample), "1180",
#                                ifelse(grepl("1344", covid_list$sample), "1344", ""))
# 
# covid_list$subset <- ifelse(grepl("CD4", covid_list$sample), "CD4",
#                                ifelse(grepl("CD8", covid_list$sample), "CD8", ""))
# covid_list$epitope <- ifelse(grepl("M24", covid_list$sample), "membrane",
#                         ifelse(grepl("S51", covid_list$sample), "spike",
#                         ifelse(grepl("S174", covid_list$sample), "spike",
#                         ifelse(grepl("S76", covid_list$sample), "spike",
#                         ifelse(grepl("ORF3a-28", covid_list$sample), "ORF3a",
#                         ifelse(grepl("NP1", covid_list$sample), "nucleoprotein",
#                         ifelse(grepl("NP13", covid_list$sample), "nucleoprotein", "")))))))
# covid_list$timepoint <- "6_month"
# covid_list$platform <- "10x_batch1"
# 
# 
# flu_list <- as.data.frame(c(d1_names$V2, d2_names$V2))
# colnames(flu_list) <- "sample"
# flu_list$patient <- ifelse(grepl("D1", flu_list$sample), "D1",
#                                ifelse(grepl("D2", flu_list$sample), "D2", ""))
# flu_list$subset <- "CD8"
# flu_list$epitope <- "flu"
# flu_list$timepoint <- ifelse(grepl("2006", flu_list$sample), "TP1",
# ifelse(grepl("2015|2016", flu_list$sample), "TP2", 
#        ifelse(grepl("NoCell", flu_list$sample), "No Cell", "")))
# flu_list$platform <- "SmartSeq2_flu"
# 
# list <- rbind(covid_list, flu_list)
# 
# combo@meta.data$"epitope" <- as.factor(list$epitope)
# combo@meta.data$"patient" <- as.factor(list$patient)
# combo@meta.data$"t-cell" <- as.factor(list$subset)
# combo@meta.data$"timepoint" <- as.factor(list$timepoint)
# combo@meta.data$"platform" <- as.factor(list$platform)
# 
# # Integration
# # split the dataset into a list by virus
# combo.list <- SplitObject(combo, split.by = "platform")
# 
# # normalize and identify variable features for each dataset independently
# combo.list <- lapply(X = combo.list, FUN = function(x) {
#     x <- NormalizeData(x)
#     x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
# })
# 
# # select features that are repeatedly variable across datasets for integration
# features <- SelectIntegrationFeatures(object.list = combo.list)
# 
# # Find anchors
# immune.anchors <- FindIntegrationAnchors(object.list = combo.list, anchor.features = features, verbose=FALSE)
# 
# # this command creates an 'integrated' data assay
# covid_flu <- IntegrateData(anchorset = immune.anchors, verbose=FALSE)
```

```{r, fig.width=16, fig.height=22}
# specify that we will perform downstream analysis on the corrected data note that the original unmodified data still resides in the 'RNA' assay
# DefaultAssay(covid_flu) <- "integrated"
# 
# # Run the standard workflow for visualization and clustering
# covid_flu <- ScaleData(covid_flu, verbose = FALSE)
# covid_flu <- RunPCA(covid_flu, npcs = 30, verbose = FALSE)
# covid_flu <- RunUMAP(covid_flu, reduction = "pca", dims = 1:30, verbose=FALSE)
# covid_flu <- FindNeighbors(covid_flu, reduction = "pca", dims = 1:30, verbose=FALSE)
# covid_flu <- FindClusters(covid_flu, resolution = 0.3)

# covid_flu <- readRDS("covid_flu_seurat_integrated.rds")
# 
# # Visualization
# DefaultAssay(covid_flu) <- "integrated"
# 
# p1 <- DimPlot(covid_flu, reduction = "umap", group.by = "seurat_clusters", label = TRUE, repel = TRUE)
# p2 <- DimPlot(covid_flu, reduction = "umap", group.by="patient")
# p3 <- DimPlot(covid_flu, reduction="umap", group.by="epitope")
# p4 <- DimPlot(covid_flu, reduction="umap", group.by="timepoint")
# p5 <- DimPlot(covid_flu, reduction="umap", group.by="t-cell")
# p6 <- DimPlot(covid_flu, reduction="pca", group.by="epitope")
# p7 <- DimPlot(covid_flu, reduction="pca", group.by="platform")
# p8 <- DimPlot(covid_flu, reduction="umap", group.by="platform")
# 
# grid.arrange(p6, p7, p1, p2, p3, p4, p5, p8, nrow=4)


# saveRDS(covid_flu, "covid_flu_seurat_integrated.rds")

```
#### Gene expression
```{r, fig.width=36, fig.height=30}
# DefaultAssay(covid_flu) <- "RNA"
# 
# VlnPlot(object = covid_flu, features=c("RPL39", "RPL30", "RPLP1", "RPS28",
#                                     "CD28", "CD3E", "PTPRC", "B2M",
#                                     "HLA-A", "HLA-B", "HLA-C", "HLA-E",
#                                     "HLA-F", "HLA-DRB1", "HLA-DPA1", "HLA-DQA1",
#                                     "CD74", "LCK", "GZMA", "CCL3", "CCL4",
#                                     "ITGB2", "ITGB7", "ITGB1", "SELPLG",
#                                     "ITGAL", "ITGA4", "CD2"),
#         ncol=4, group.by = "seurat_clusters")


```

```{r, fig.width=25, fig.height=40}
# DefaultAssay(covid_flu) <- "RNA"
# 
# VlnPlot(object = covid_flu, features=c("CD27", "CD28", "CD69", "CD38",
#                                       "PTPRC",   "CCR7", "TCF7", "CD44",
#                                        "IL7R", "S100A4", "SELL", "HLA-DRB1",
#                                        "TBX21", "IFNG", "LTA", "TNF", 
#                                        "CISH", "IL2RA",  "IL13", "TGFB1",
#                                        "CCL3", "CCL4", "CCL5", "CCR1", 
#                                        "CCR2",  "CCR4", "CCR5", 
#                                        "CXCR3", "CXCR4", "CXCR6", 
#                                        "PRF1", "GNLY", "NKG7", "XCL1", 
#                                        "GZMA", "GZMB", "GZMH", "GZMK",
#                                        "OAS1", "MX1", "ISG15",
#                                        "IFI44L", "MKI67", "TUBA1B",
#                                        "HAVCR2", "CD244", "KLRG1", "TIGIT"),
#         ncol=4, group.by = "seurat_clusters")

```

```{r, fig.width=20, fig.height=40}
# DefaultAssay(covid_flu) <- "RNA"
# 
# VlnPlot(object = covid_flu, features=c("CD27", "CD28", "CD69", "CD38",
#                                       "PTPRC",   "CCR7", "TCF7", "CD44",
#                                        "IL7R", "S100A4", "SELL", "HLA-DRB1",
#                                        "TBX21", "IFNG", "LTA", "TNF", 
#                                        "CISH", "IL2RA",  "IL13", "TGFB1",
#                                        "CCL3", "CCL4", "CCL5", "CCR1", 
#                                        "CCR2",  "CCR4", "CCR5", 
#                                        "CXCR3", "CXCR4", "CXCR6", 
#                                        "PRF1", "GNLY", "NKG7", "XCL1", 
#                                        "GZMA", "GZMB", "GZMH", "GZMK",
#                                        "OAS1", "MX1", "ISG15",
#                                        "IFI44L", "MKI67", "TUBA1B",
#                                        "HAVCR2", "CD244", "KLRG1", "TIGIT"),
#         ncol=4, group.by = "epitope")

```

#### Cluster markers
```{r}
# covid_flu.markers <- FindAllMarkers(object = covid_flu, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
# change only.pos=FALSE to get up/down regulation?

# cluster_0 <- covid_flu.markers[covid_flu.markers$cluster == "0", ]
# cluster_1 <- covid_flu.markers[covid_flu.markers$cluster == "1", ]
# cluster_2 <- covid_flu.markers[covid_flu.markers$cluster == "2", ]
# cluster_3 <- covid_flu.markers[covid_flu.markers$cluster == "3", ]
# cluster_4 <- covid_flu.markers[covid_flu.markers$cluster == "4", ]
# cluster_5 <- covid_flu.markers[covid_flu.markers$cluster == "5", ]
# cluster_6 <- covid_flu.markers[covid_flu.markers$cluster == "6", ]
# cluster_7 <- covid_flu.markers[covid_flu.markers$cluster == "7", ]
# cluster_8 <- covid_flu.markers[covid_flu.markers$cluster == "8", ]
# cluster_9 <- covid_flu.markers[covid_flu.markers$cluster == "9", ]
# cluster_10 <- covid_flu.markers[covid_flu.markers$cluster == "10", ]
# cluster_11 <- covid_flu.markers[covid_flu.markers$cluster == "11", ]
# 
# setwd('/t1-data/user/lfelce/scRNA-Seq/SmartSeq2_T-cells/')
# tmarkers <- read.csv("Markers_for_Ling.csv", header=F)
# 
# genes_0 <- tmarkers[is.element(tmarkers$V1, cluster_0$gene),]
# genes_1 <- tmarkers[is.element(tmarkers$V1, cluster_1$gene),]
# genes_2 <- tmarkers[is.element(tmarkers$V1, cluster_2$gene),]
# genes_3 <- tmarkers[is.element(tmarkers$V1, cluster_3$gene),]
# genes_4 <- tmarkers[is.element(tmarkers$V1, cluster_4$gene),]
# genes_5 <- tmarkers[is.element(tmarkers$V1, cluster_5$gene),]
# genes_6 <- tmarkers[is.element(tmarkers$V1, cluster_6$gene),]
# genes_7 <- tmarkers[is.element(tmarkers$V1, cluster_7$gene),]
# genes_8 <- tmarkers[is.element(tmarkers$V1, cluster_8$gene),]
# genes_9 <- tmarkers[is.element(tmarkers$V1, cluster_9$gene),]
# genes_10 <- tmarkers[is.element(tmarkers$V1, cluster_10$gene),]
# genes_11 <- tmarkers[is.element(tmarkers$V1, cluster_11$gene),]

# cluster0.markers <- FindMarkers(covid_flu, ident.1 = 0, ident.2 = c(1:15), min.pct = 0.25)
```

### Subset CD8 cells only from Covid Flu integrated object
```{r, fig.height=18, fig.width=15}
# covid_flu <- readRDS("covid_flu_seurat_integrated.rds")
# DefaultAssay(covid_flu) <- "integrated"
# Idents(covid_flu) <- "t-cell"
# cd8 <- subset(covid_flu, idents = "CD8")
# 
# # specify that we will perform downstream analysis on the corrected data note that the original unmodified data still resides in the 'RNA' assay
# 
# DefaultAssay(cd8) <- "integrated"
# 
# # Run the standard workflow for visualization and clustering
# cd8 <- ScaleData(cd8, verbose = FALSE)
# cd8 <- RunPCA(cd8, npcs = 30, verbose = FALSE)
# cd8 <- RunUMAP(cd8, reduction = "pca", dims = 1:25, verbose=FALSE)
# cd8 <- FindNeighbors(cd8, reduction = "pca", dims = 1:25, verbose=FALSE)
# cd8 <- FindClusters(cd8, resolution = 0.3)
# 
# # Visualization
# DefaultAssay(cd8) <- "integrated"
# p1 <- DimPlot(cd8, reduction = "umap", group.by = "seurat_clusters", label = TRUE, repel = TRUE)
# p2 <- DimPlot(cd8, reduction = "umap", group.by="patient")
# p3 <- DimPlot(cd8, reduction="umap", group.by="epitope")
# p4 <- DimPlot(cd8, reduction="umap", group.by="timepoint")
# p5 <- DimPlot(cd8, reduction="umap", group.by="t-cell")
# p6 <- DimPlot(cd8, reduction="pca", group.by="epitope")
# p7 <- DimPlot(cd8, reduction="pca", group.by="platform")
# p8 <- DimPlot(cd8, reduction="umap", group.by="platform")
# 
# grid.arrange(p6, p7, p1, p2, p3, p4, p5, p8, nrow=4)
# 
# saveRDS(cd8, "covid_flu_cd8_seurat_integrated.rds")
```

#### Cluster markers
```{r}
# cd8.markers <- FindAllMarkers(object = cd8, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
# # change only.pos=FALSE to get up/down regulation?
# 
# cluster_0 <- cd8.markers[cd8.markers$cluster == "0", ]
# cluster_1 <- cd8.markers[cd8.markers$cluster == "1", ]
# cluster_2 <- cd8.markers[cd8.markers$cluster == "2", ]
# cluster_3 <- cd8.markers[cd8.markers$cluster == "3", ]
# cluster_4 <- cd8.markers[cd8.markers$cluster == "4", ]
# cluster_5 <- cd8.markers[cd8.markers$cluster == "5", ]
# cluster_6 <- cd8.markers[cd8.markers$cluster == "6", ]
# cluster_7 <- cd8.markers[cd8.markers$cluster == "7", ]
# cluster_8 <- cd8.markers[cd8.markers$cluster == "8", ]
# cluster_9 <- cd8.markers[cd8.markers$cluster == "9", ]
# cluster_10 <- cd8.markers[cd8.markers$cluster == "10", ]
# 
# setwd('/t1-data/user/lfelce/scRNA-Seq/SmartSeq2_T-cells/')
# tmarkers <- read.csv("Markers_for_Ling.csv", header=F)
#  
# genes_0 <- tmarkers[is.element(tmarkers$V1, cluster_0$gene),]
# genes_1 <- tmarkers[is.element(tmarkers$V1, cluster_1$gene),]
# genes_2 <- tmarkers[is.element(tmarkers$V1, cluster_2$gene),]
# genes_3 <- tmarkers[is.element(tmarkers$V1, cluster_3$gene),]
# genes_4 <- tmarkers[is.element(tmarkers$V1, cluster_4$gene),]
# genes_5 <- tmarkers[is.element(tmarkers$V1, cluster_5$gene),]
# genes_6 <- tmarkers[is.element(tmarkers$V1, cluster_6$gene),]
# genes_7 <- tmarkers[is.element(tmarkers$V1, cluster_7$gene),]
# genes_8 <- tmarkers[is.element(tmarkers$V1, cluster_8$gene),]
# genes_9 <- tmarkers[is.element(tmarkers$V1, cluster_9$gene),]
# genes_10 <- tmarkers[is.element(tmarkers$V1, cluster_10$gene),]

# cluster0.markers <- FindMarkers(cd8, ident.1 = 0, ident.2 = c(1:15), min.pct = 0.25)
```

#### Gene expression
```{r, fig.width=16, fig.height=40}
# DefaultAssay(cd8) <- "RNA"
# 
# VlnPlot(object = cd8, features=c("CD27", "CD28", "CD69", "CD38",
#                                       "PTPRC",   "CCR7", "TCF7", "CD44",
#                                        "IL7R", "S100A4", "SELL", "HLA-DRB1",
#                                        "TBX21", "IFNG", "LTA", "TNF", 
#                                        "CISH", "IL2RA",  "IL13", "TGFB1",
#                                        "CCL3", "CCL4", "CCL5", "CCR1", 
#                                        "CCR2",  "CCR4", "CCR5", 
#                                        "CXCR3", "CXCR4", "CXCR6", 
#                                        "PRF1", "GNLY", "NKG7", "XCL1", 
#                                        "GZMA", "GZMB", "GZMH", "GZMK",
#                                        "OAS1", "MX1", "ISG15",
#                                        "IFI44L", "MKI67", "TUBA1B",
#                                        "HAVCR2", "CD244", "KLRG1", "TIGIT"),
#         ncol=4, group.by = "epitope")
```
#### Gene expression
```{r, fig.width=24, fig.height=40}
# DefaultAssay(cd8) <- "RNA"
# 
# VlnPlot(object = cd8, features=c("CD27", "CD28", "CD69", "CD38",
#                                       "PTPRC",   "CCR7", "TCF7", "CD44",
#                                        "IL7R", "S100A4", "SELL", "HLA-DRB1",
#                                        "TBX21", "IFNG", "LTA", "TNF", 
#                                        "CISH", "IL2RA",  "IL13", "TGFB1",
#                                        "CCL3", "CCL4", "CCL5", "CCR1", 
#                                        "CCR2",  "CCR4", "CCR5", 
#                                        "CXCR3", "CXCR4", "CXCR6", 
#                                        "PRF1", "GNLY", "NKG7", "XCL1", 
#                                        "GZMA", "GZMB", "GZMH", "GZMK",
#                                        "OAS1", "MX1", "ISG15",
#                                        "IFI44L", "MKI67", "TUBA1B",
#                                        "HAVCR2", "CD244", "KLRG1", "TIGIT"),
#         ncol=4, group.by = "seurat_clusters")
```
### Subset CD8 cells only from Covid object
```{r, fig.height=8, fig.width=14}
# read in batch1 seurat and subset CD8 cells
batch1 <- readRDS("batch1_seurat.rds")
DefaultAssay(batch1) <- "RNA"
Idents(batch1) <- "t-cell"
cd8 <- subset(batch1, idents = "CD8")

# Run the standard workflow for visualization and clustering
cd8 <- ScaleData(cd8, verbose = FALSE)
cd8 <- RunPCA(cd8, npcs = 30, verbose = FALSE)
cd8 <- RunUMAP(cd8, reduction = "pca", dims = 1:25, verbose=FALSE)
cd8 <- FindNeighbors(cd8, reduction = "pca", dims = 1:25, verbose=FALSE)
cd8 <- FindClusters(cd8, resolution = 0.5)
# Visualization
p1 <- DimPlot(cd8, reduction = "umap", group.by = "seurat_clusters", label = TRUE, repel = TRUE)
p2 <- DimPlot(cd8, reduction = "umap", group.by="patient")
p3 <- DimPlot(cd8, reduction="umap", group.by="epitope")
p4 <- DimPlot(cd8, reduction="pca", group.by="epitope")

grid.arrange(p1, p2, p3, p4, nrow=2)
saveRDS(cd8, "batch1_cd8_seurat.rds")
```
#### Gene expression by epitope
```{r, fig.width=20, fig.height=26}
VlnPlot(object = cd8, features=c("PTPRC", "CCR7", "SELL", "TCF7", "IL7R", "S100A4", 
                                  "CD44","CD27", "CD28", "CD69", "CD38","HLA-DRB1",
                                  "TBX21", "IFNG", "LTA", "TNF", "CISH", "IL2RA",
                                  "IL13", "TGFB1", "CCL3", "CCL4", "CCL5", "CCR1",
                                  "CCR2",  "CCR4", "CCR5", "CXCR3", "CXCR4", "CXCR6", 
                                  "GZMA", "GZMB", "GZMH", "GZMK", "PRF1", "GNLY",
                                   "NKG7", "XCL1", "FASLG", "OAS1", "MX1", "ISG15",  
                                  "MKI67", "TUBA1B","HAVCR2", "CD244", "KLRG1", "TIGIT"),
        ncol=6, group.by = "epitope")
```
#### Gene expression by cluster
```{r, fig.width=26, fig.height=26}
VlnPlot(object = cd8, features=c("PTPRC", "CCR7", "SELL", "TCF7", "IL7R", "S100A4", 
                                  "CD44","CD27", "CD28", "CD69", "CD38","HLA-DRB1",
                                  "TBX21", "IFNG", "LTA", "TNF", "CISH", "IL2RA",
                                  "IL13", "TGFB1", "CCL3", "CCL4", "CCL5", "CCR1",
                                  "CCR2",  "CCR4", "CCR5", "CXCR3", "CXCR4", "CXCR6", 
                                  "GZMA", "GZMB", "GZMH", "GZMK", "PRF1", "GNLY",
                                   "NKG7", "XCL1", "FASLG", "OAS1", "MX1", "ISG15",  
                                  "MKI67", "TUBA1B","HAVCR2", "CD244", "KLRG1", "TIGIT"),
        ncol=6, group.by = "seurat_clusters")
```
### Subset spike cells only from Covid object
```{r, fig.height=7, fig.width=12}
# read in batch1 seurat and subset CD8 cells
batch1 <- readRDS("batch1_seurat.rds")
DefaultAssay(batch1) <- "RNA"

Idents(batch1) <- "epitope"
spike <- subset(batch1, idents = "spike")

# Run the standard workflow for visualization and clustering
spike <- ScaleData(spike, verbose = FALSE)
spike <- RunPCA(spike, npcs = 30, verbose = FALSE)
spike <- RunUMAP(spike, reduction = "pca", dims = 1:25, verbose=FALSE)
spike <- FindNeighbors(spike, reduction = "pca", dims = 1:25, verbose=FALSE)
spike <- FindClusters(spike, resolution = 0.3)
# Visualization
p1 <- DimPlot(spike, reduction = "umap", group.by = "seurat_clusters", label = TRUE, repel = TRUE)
p2 <- DimPlot(spike, reduction = "umap", group.by="hashtag_classification")
p3 <- DimPlot(spike, reduction="umap", group.by="t-cell")
p4 <- DimPlot(spike, reduction="pca", group.by="patient")
grid.arrange(p1, p2, p3, p4, nrow=2)

saveRDS(spike, "batch1_spike_seurat.rds")
```
#### Gene expression by sample
```{r, fig.width=20, fig.height=26}
VlnPlot(object = spike, features=c("PTPRC", "CCR7", "SELL", "TCF7", "IL7R", "S100A4", 
                                  "CD44","CD27", "CD28", "CD69", "CD38","HLA-DRB1",
                                  "TBX21", "IFNG", "LTA", "TNF", "CISH", "IL2RA",
                                  "IL13", "TGFB1", "CCL3", "CCL4", "CCL5", "CCR1",
                                  "CCR2",  "CCR4", "CCR5", "CXCR3", "CXCR4", "CXCR6", 
                                  "GZMA", "GZMB", "GZMH", "GZMK", "PRF1", "GNLY",
                                   "NKG7", "XCL1", "FASLG", "OAS1", "MX1", "ISG15",  
                                  "MKI67", "TUBA1B","HAVCR2", "CD244", "KLRG1", "TIGIT"),
        ncol=6, group.by = "hashtag_classification")
```
#### Gene expression by cluster
```{r, fig.width=26, fig.height=26}
VlnPlot(object = spike, features=c("PTPRC", "CCR7", "SELL", "TCF7", "IL7R", "S100A4", 
                                  "CD44","CD27", "CD28", "CD69", "CD38","HLA-DRB1",
                                  "TBX21", "IFNG", "LTA", "TNF", "CISH", "IL2RA",
                                  "IL13", "TGFB1", "CCL3", "CCL4", "CCL5", "CCR1",
                                  "CCR2",  "CCR4", "CCR5", "CXCR3", "CXCR4", "CXCR6", 
                                  "GZMA", "GZMB", "GZMH", "GZMK", "PRF1", "GNLY",
                                   "NKG7", "XCL1", "FASLG", "OAS1", "MX1", "ISG15",  
                                  "MKI67", "TUBA1B","HAVCR2", "CD244", "KLRG1", "TIGIT"),
        ncol=6, group.by = "seurat_clusters")
```



### Combined GEX and Antibody Hashtag
Combined by putting two rows in libraries.csv for Feature Barcode analysis (cellranger count) for library_type Gene Expression and Antibody Capture. Followed 
https://satijalab.org/seurat/articles/multimodal_vignette.html to create Seurat object with two assays - RNA and HTO. Then use HTODemux().
```{r}
# # Read in data
# batch1.counts <- Read10X(data.dir = "/t1-data/user/lfelce/10x_DONG171220/counts_combined_2/filtered_feature_bc_matrix")
# 
# # Remove -1 from barcodes
# colnames(batch1.counts$`Gene Expression`) <- gsub(x = colnames(batch1.counts$`Gene Expression`), pattern = "-1", replacement = "")
# 
# colnames(batch1.counts$`Antibody Capture`) <- gsub(x = colnames(batch1.counts$`Antibody Capture`), pattern = "-1", replacement = "")
# 
# # Remove Hashtag 3 from Antibody Capture (no cells)
# batch1.counts$`Antibody Capture` <- batch1.counts$`Antibody Capture`[-3,]
# 
# # Rename rows to sample names
# rownames(batch1.counts$`Antibody Capture`) <- c("1180_CD4_M24", "1344_CD4_S51", "1344_CD4_S174", "1180_CD8_S76", "1180_CD8_ORF3a-28", "1344_CD8_NP1", "1344_CD8_NP13")
# 
# # Create Seurat object from Gene Expression
# batch1 <- CreateSeuratObject(counts = batch1.counts$`Gene Expression`)
# 
# # Normalize RNA data with log normalization
# DefaultAssay(batch1) <- "RNA"
# batch1 <- NormalizeData(batch1)
# 
# # Find and scale variable features
# batch1 <- FindVariableFeatures(batch1, selection.method = "mean.var.plot")
# batch1 <- ScaleData(batch1, features = VariableFeatures(batch1))
# 
# # Add HTO data as a new assay independent from RNA
# batch1[["HTO"]] <- CreateAssayObject(counts = batch1.counts$`Antibody Capture`)
# 
# # Normalize HTO data, here we use centered log-ratio (CLR) transformation
# batch1 <- NormalizeData(batch1, assay = "HTO", normalization.method = "CLR")
# 
# # Demultiplex cells based on hashtag enrichment - use HTODemux() to assign single cells back to sample origins
# # If you have a very large dataset we suggest using k_function = 'clara'. This is a k-medoid
# # clustering function for large applications You can also play with additional parameters (see
# # documentation for HTODemux()) to adjust the threshold for classification Here we are using the
# # default settings
# batch1 <- HTODemux(batch1, assay = "HTO", positive.quantile = 0.99)
# 
# table(batch1$HTO_classification.global)
# 
# Idents(batch1) <- "HTO_classification.global"
# VlnPlot(batch1, features = "nCount_RNA", pt.size = 0.1, log = TRUE)

```
Cluster and visualize cells using the usual scRNA-seq workflow, and examine for the potential presence of batch effects.

```{r, fig.width=18, fig.height=16}
# # Extract the singlets
# batch1.singlet <- subset(batch1, idents = "Singlet")
# 
# # Assign metadata to object
# patient_list <- as.data.frame(batch1.singlet@meta.data$HTO_classification)
# colnames(patient_list) <- "sample"
# patient_list$patient <- ifelse(grepl("1180", patient_list$sample), "1180",
#                                ifelse(grepl("1344", patient_list$sample), "1344", ""))
# patient_list$subset <- ifelse(grepl("CD4", patient_list$sample), "CD4",
#                                ifelse(grepl("CD8", patient_list$sample), "CD8", ""))
# patient_list$epitope <- ifelse(grepl("M24", patient_list$sample), "membrane",
#                         ifelse(grepl("S51", patient_list$sample), "spike",
#                         ifelse(grepl("S174", patient_list$sample), "spike",
#                         ifelse(grepl("S76", patient_list$sample), "spike",
#                         ifelse(grepl("ORF3a-28", patient_list$sample), "ORF3a",
#                         ifelse(grepl("NP1", patient_list$sample), "nucleoprotein",
#                         ifelse(grepl("NP13", patient_list$sample), "nucleoprotein", "")))))))
# 
# batch1.singlet@meta.data$"epitope" <- as.factor(patient_list$epitope)
# batch1.singlet@meta.data$"patient" <- as.factor(patient_list$patient)
# batch1.singlet@meta.data$"t-cell" <- as.factor(patient_list$subset)
# 
# # Select the top 1000 most variable features
# batch1.singlet <- FindVariableFeatures(batch1.singlet, selection.method = "mean.var.plot")
# 
# # Scaling RNA data
# batch1.singlet <- ScaleData(batch1.singlet, features = VariableFeatures(batch1.singlet))
# 
# # Run PCA
# batch1.singlet <- RunPCA(batch1.singlet, features = VariableFeatures(batch1.singlet))
# 
# # Select top PCs for clustering
# batch1.singlet <- FindNeighbors(batch1.singlet, reduction = "pca", dims = 1:20)
# batch1.singlet <- FindClusters(batch1.singlet, resolution = 0.3, verbose = FALSE)
# batch1.singlet <- RunUMAP(batch1.singlet, dims = 1:20)
# batch1.singlet <- RunTSNE(batch1.singlet, reduction = "pca", dims = 1:20)
# 
# # Projecting singlet identities on UMAP visualization
# tsne <- DimPlot(batch1.singlet, reduction="tsne", group.by = "HTO_classification")
# pca <- DimPlot(batch1.singlet, reduction="pca", group.by="HTO_classification")
# 
# umap_cluster <- DimPlot(batch1.singlet, reduction="umap", group.by = "seurat_clusters")
# umap_tcell <- DimPlot(batch1.singlet, reduction="umap", group.by = "t-cell")
# umap_epitope <- DimPlot(batch1.singlet, reduction="umap", group.by = "epitope")
# umap_patient <- DimPlot(batch1.singlet, reduction="umap", group.by = "patient")
# 
# grid.arrange(tsne, pca, umap_cluster, umap_tcell, umap_epitope, umap_patient, nrow=3)

```
