---
title: "SARS-CoV-2 6 month T cells"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(plyr)
library(dplyr)
library(patchwork)
library(cowplot)
library(ggplot2)
library(data.table)
library(tidyverse)
library(harmony)
library(gridExtra)
```

# 10x single cell RNA sequencing from SARS-CoV-2 CD4/8 tetramer sorted T cells

## Batch 1
Dong 171220 = batch1
Hashtag Ab 1 1180_6Mon, expanded, DR0101_M24 tetramer, CD4
Hashtag Ab 2 1344_6Mon, expanded, DR1501_S51 tetramer, CD4
Hashtag Ab 4 1344_6Mon, expanded, DR1501_S174 tetramer, CD4
Hashtag Ab 5 1180_6Mon, expanded, A3_S76 tetramer, CD8
Hashtag Ab 6 1180_6Mon, expanded, A1-ORF3a-28 tetramer, CD8
Hashtag Ab 7 1344_6Mon, expanded, B27_NP1 tetramer, CD8
Hashtag Ab 8 1344_6Mon, expanded, B27_NP13 tetramer, CD8


### Combined GEX and Antibody Hashtag
Combined by putting two rows in libraries.csv for Feature Barcode analysis (cellranger count) for library_type Gene Expression and Antibody Capture. Followed 
https://satijalab.org/seurat/articles/multimodal_vignette.html to create Seurat object with two assays - RNA and HTO. Then use HTODemux().

Discussion with Andrew Kwok - decided this was the better method as Cellranger already intersects barcodes, so retains more cells for analysis (4656 singlets vs 3695)
```{r}
# Read in data
batch1.counts <- Read10X(data.dir = "/t1-data/user/lfelce/10x_DONG171220/counts_combined_2/filtered_feature_bc_matrix")

# Remove -1 from barcodes
colnames(batch1.counts$`Gene Expression`) <- gsub(x = colnames(batch1.counts$`Gene Expression`), pattern = "-1", replacement = "")

colnames(batch1.counts$`Antibody Capture`) <- gsub(x = colnames(batch1.counts$`Antibody Capture`), pattern = "-1", replacement = "")

# Remove Hashtag 3 from Antibody Capture (no cells)
batch1.counts$`Antibody Capture` <- batch1.counts$`Antibody Capture`[-3,]

# Rename rows to sample names
rownames(batch1.counts$`Antibody Capture`) <- c("1180_CD4_M24", "1344_CD4_S51", "1344_CD4_S174", "1180_CD8_S76", "1180_CD8_ORF3a-28", "1344_CD8_NP1", "1344_CD8_NP13")

# Create Seurat object from Gene Expression
batch1 <- CreateSeuratObject(counts = batch1.counts$`Gene Expression`)

# Normalize RNA data with log normalization
DefaultAssay(batch1) <- "RNA"
batch1 <- NormalizeData(batch1)

# Find and scale variable features
batch1 <- FindVariableFeatures(batch1, selection.method = "mean.var.plot")
batch1 <- ScaleData(batch1, features = VariableFeatures(batch1))

# Add HTO data as a new assay independent from RNA
batch1[["HTO"]] <- CreateAssayObject(counts = batch1.counts$`Antibody Capture`)

# Normalize HTO data, here we use centered log-ratio (CLR) transformation
batch1 <- NormalizeData(batch1, assay = "HTO", normalization.method = "CLR")

# Demultiplex cells based on hashtag enrichment - use HTODemux() to assign single cells back to sample origins
# If you have a very large dataset we suggest using k_function = 'clara'. This is a k-medoid
# clustering function for large applications You can also play with additional parameters (see
# documentation for HTODemux()) to adjust the threshold for classification Here we are using the
# default settings
batch1 <- HTODemux(batch1, assay = "HTO", positive.quantile = 0.99)

table(batch1$HTO_classification.global)

Idents(batch1) <- "HTO_classification.global"
VlnPlot(batch1, features = "nCount_RNA", pt.size = 0.1, log = TRUE)

```
Cluster and visualize cells using the usual scRNA-seq workflow, and examine for the potential presence of batch effects.

```{r}
# Extract the singlets
# batch1.singlet <- subset(batch1, idents = "Singlet")
# 
# # Assign metadata to object
# patient_list <- as.data.frame(batch1.singlet@meta.data$HTO_classification)
# colnames(patient_list) <- "sample"
# patient_list$patient <- ifelse(grepl("1180", patient_list$sample), "1180",
#                                ifelse(grepl("1344", patient_list$sample), "1344", ""))
# patient_list$subset <- ifelse(grepl("CD4", patient_list$sample), "CD4",
#                                ifelse(grepl("CD8", patient_list$sample), "CD8", ""))
# patient_list$epitope <- ifelse(grepl("M24", patient_list$sample), "membrane",
#                         ifelse(grepl("S51", patient_list$sample), "spike",
#                         ifelse(grepl("S174", patient_list$sample), "spike",
#                         ifelse(grepl("S76", patient_list$sample), "spike",
#                         ifelse(grepl("ORF3a-28", patient_list$sample), "ORF3a",
#                         ifelse(grepl("NP1", patient_list$sample), "nucleoprotein",
#                         ifelse(grepl("NP13", patient_list$sample), "nucleoprotein", "")))))))
# 
# batch1.singlet@meta.data$"epitope" <- as.factor(patient_list$epitope)
# batch1.singlet@meta.data$"patient" <- as.factor(patient_list$patient)
# batch1.singlet@meta.data$"t-cell" <- as.factor(patient_list$subset)

```


```{r qc, fig.height=5, fig.width=16}
batch1.singlet <- readRDS("batch1_combined_seurat.rds")

Idents(batch1.singlet) <- "HTO_classification"

# non-UMI data - nUMI represents sum of non-normalised values within cells
# calculate % mitochondrial genes and store in separate object
# use object@raw.data represents non-transformed and non-log-normalised counts
# % of UMI mapping to MT-genes is a common scRNA-seq QC metric
mito.genes <- grep(pattern = "^MT-", x = rownames(batch1.singlet@assays[["RNA"]]), value = TRUE)

percent.mito <- Matrix::colSums(batch1.singlet@assays[["RNA"]][mito.genes, ])/Matrix::colSums(batch1.singlet@assays[["RNA"]])

batch1.singlet <- AddMetaData(object = batch1.singlet, metadata = percent.mito, col.name = "percent.mito")

VlnPlot(object = batch1.singlet, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), ncol = 3)

```

```{r, fig.height=4, fig.width=6}
# can use GenePlot to visualise gene-gene relationships and anything else calculated by object
FeatureScatter(object=batch1.singlet, feature1="nCount_RNA", feature2="nFeature_RNA")
```


```{r filtering}
# Filter out cells that have unique gene counts (nFeature_RNA) over 4000 or less than 200 
# Note that > and < are used to define a'gate'.  
# -Inf and Inf should be used if don't want a lower or upper threshold.
batch1.singlet <- subset(x = batch1.singlet, subset = nFeature_RNA > 200 & nFeature_RNA < 4500 & percent.mito >  -Inf & percent.mito < 0.05 )
```


```{r}
# After removing unwanted cells from dataset, need to normalise data. Use global-scaling normalisation method "LogNormalize" - normalises gene expression measurements for each cell by total expression, multiplies by scale factor (10,000 default) and log-tranforms result.
batch1.singlet <- NormalizeData(object = batch1.singlet, normalization.method = "LogNormalize", scale.factor = 10000)

# Select the top 1000 most variable features
batch1.singlet <- FindVariableFeatures(batch1.singlet, selection.method = "mean.var.plot")

# Scaling RNA data
batch1.singlet <- ScaleData(batch1.singlet, features = VariableFeatures(batch1.singlet))

# Run PCA
batch1.singlet <- RunPCA(batch1.singlet, features = VariableFeatures(batch1.singlet))

# Select top PCs for clustering
batch1.singlet <- FindNeighbors(batch1.singlet, reduction = "pca", dims = 1:30, verbose=FALSE)
batch1.singlet <- FindClusters(batch1.singlet, resolution = 0.5, verbose = FALSE)
batch1.singlet <- RunUMAP(batch1.singlet, dims = 1:30, verbose=FALSE)
batch1.singlet <- RunTSNE(batch1.singlet, reduction = "pca", dims = 1:30, verbose=FALSE)

# Projecting singlet identities on UMAP visualization
tsne <- DimPlot(batch1.singlet, reduction="tsne", group.by = "epitope")
pca <- DimPlot(batch1.singlet, reduction="pca", group.by="epitope")

umap_cluster <- DimPlot(batch1.singlet, reduction="umap", group.by = "seurat_clusters")
umap_tcell <- DimPlot(batch1.singlet, reduction="umap", group.by = "t-cell")
umap_epitope <- DimPlot(batch1.singlet, reduction="umap", group.by = "epitope")
umap_patient <- DimPlot(batch1.singlet, reduction="umap", group.by = "HTO_classification")
```

```{r, fig.width=16, fig.height=16}

grid.arrange(tsne, pca, umap_cluster, umap_tcell, umap_epitope, umap_patient, nrow=3)

```

```{r, fig.width=10, fig.height=7}
FeaturePlot(object = batch1.singlet, features = c("CD4", "CD8A", "CD8B" ), cols = c("grey", "blue"), reduction = "umap")
```



#### Gene expression (violin plots)
```{r, fig.width=18, fig.height=24}
# batch1 <- readRDS("batch1_combined_seurat.rds")

VlnPlot(object = batch1.singlet, features=c("PTPRC", "CCR7", "SELL", "TCF7", "IL7R", "S100A4",
                                  "CD44","CD27", "CD28", "CD69", "CD38","HLA-DRB1",
                                  "IFNG", "CCL4", "CCL5", "CCR4", "CXCR3", "CXCR4",                                                   "SELPLG", "ITGA4", "ITGAL", "ITGB2", "ITGAE", "ICAM1",
                                  "GZMA", "GZMB", "GZMH", "GZMK", "PRF1", "GNLY",
                                  "NKG7", "FASLG", "FAS",  "ISG15", "MX1", "OAS1",
                                  "TUBA1B","LAG3","PDCD1", "HAVCR2", "TIGIT", "CTLA4",
                                 "KLRD1", "FGFBP2", "KLRC1", "KLRC3", "KLRB1", "KLRC2"),
        ncol=6, group.by = "epitope")
```
```{r, fig.width=32, fig.height=24}
# batch1 <- readRDS("batch1_combined_seurat.rds")

VlnPlot(object = batch1.singlet, features=c("PTPRC", "CCR7", "SELL", "TCF7", "IL7R", "S100A4",
                                  "CD44","CD27", "CD28", "CD69", "CD38","HLA-DRB1",
                                  "IFNG", "CCL4", "CCL5", "CCR4", "CXCR3", "CXCR4",                                                   "SELPLG", "ITGA4", "ITGAL", "ITGB2", "ITGAE", "ICAM1",
                                  "GZMA", "GZMB", "GZMH", "GZMK", "PRF1", "GNLY",
                                  "NKG7", "FASLG", "FAS",  "ISG15", "MX1", "OAS1",
                                  "TUBA1B","LAG3","PDCD1", "HAVCR2", "TIGIT", "CTLA4",
                                 "KLRD1", "FGFBP2", "KLRC1", "KLRC3", "KLRB1", "KLRC2"),
        ncol=6, group.by = "seurat_clusters")
```
### Subset CD8 cells only from Covid object
```{r, fig.height=12, fig.width=14}
# read in batch1 seurat and subset CD8 cells
# batch1 <- readRDS("batch1_combined_seurat.rds")
DefaultAssay(batch1.singlet) <- "RNA"
Idents(batch1.singlet) <- "t-cell"
cd8 <- subset(batch1.singlet, idents = "CD8")

# Run the standard workflow for visualization and clustering
cd8 <- ScaleData(cd8, verbose = FALSE)
cd8 <- RunPCA(cd8, npcs = 30, verbose = FALSE)
cd8 <- RunUMAP(cd8, reduction = "pca", dims = 1:30, verbose=FALSE)
cd8 <- FindNeighbors(cd8, reduction = "pca", dims = 1:30, verbose=FALSE)
cd8 <- FindClusters(cd8, resolution = 0.5)
# Visualization
p1 <- DimPlot(cd8, reduction = "umap", group.by = "seurat_clusters", label = TRUE, repel = TRUE)
p2 <- DimPlot(cd8, reduction = "umap", group.by="patient")
p3 <- DimPlot(cd8, reduction="umap", group.by="epitope")
p4 <- DimPlot(cd8, reduction="pca", group.by="epitope")
p5 <- DimPlot(cd8, reduction="umap", group.by="HTO_classification")

grid.arrange(p1, p2, p3, p5, p4, nrow=3)

# saveRDS(cd8, "batch1_combined_cd8_seurat.rds")
```
#### Gene expression 
```{r, fig.width=26, fig.height=24}
# cd8 <- readRDS("batch1_combined_cd8_seurat.rds")
VlnPlot(object = cd8, features=c("PTPRC", "CCR7", "SELL", "TCF7", "IL7R", "S100A4",
                                  "CD44","CD27", "CD28", "CD69", "CD38","HLA-DRB1",
                                  "IFNG", "CCL4", "CCL5", "CCR4", "CXCR3", "CXCR4",                                                   "SELPLG", "ITGA4", "ITGAL", "ITGB2", "ITGAE", "ICAM1",
                                  "GZMA", "GZMB", "GZMH", "GZMK", "PRF1", "GNLY",
                                  "NKG7", "FASLG", "FAS",  "ISG15", "MX1", "OAS1",
                                  "TUBA1B","LAG3","PDCD1", "HAVCR2", "TIGIT", "CTLA4",
                                 "KLRD1", "FGFBP2", "KLRC1", "KLRC3", "KLRB1", "KLRC2"),
        ncol=6, group.by = "seurat_clusters")
```

```{r, fig.width=22, fig.height=28}
VlnPlot(object = cd8, features=c("PTPRC", "CCR7", "SELL", "TCF7", "IL7R", "S100A4",
                                  "CD44","CD27", "CD28", "CD69", "CD38","HLA-DRB1",
                                  "IFNG", "CCL4", "CCL5", "CCR4", "CXCR3", "CXCR4",                                                   "SELPLG", "ITGA4", "ITGAL", "ITGB2", "ITGAE", "ICAM1",
                                  "GZMA", "GZMB", "GZMH", "GZMK", "PRF1", "GNLY",
                                  "NKG7", "FASLG", "FAS",  "ISG15", "MX1", "OAS1",
                                  "TUBA1B","LAG3","PDCD1", "HAVCR2", "TIGIT", "CTLA4",
                                 "KLRD1", "FGFBP2", "KLRC1", "KLRC3", "KLRB1", "KLRC2"),
        ncol=6, group.by = "epitope")
```
```{r, fig.width=10, fig.height=7}
FeaturePlot(object = cd8, features = c("CD4", "CD8A", "CD8B" ), cols = c("grey", "blue"), reduction = "umap")

```

### Subset spike cells only from Covid object
```{r, fig.height=8, fig.width=13}
# read in batch1 seurat and subset CD8 cells
# batch1 <- readRDS("batch1_combined_seurat.rds")
DefaultAssay(batch1.singlet) <- "RNA"

Idents(batch1.singlet) <- "epitope"
spike <- subset(batch1.singlet, idents = "spike")

# Run the standard workflow for visualization and clustering
spike <- ScaleData(spike, verbose = FALSE)
spike <- RunPCA(spike, npcs = 30, verbose = FALSE)
spike <- RunUMAP(spike, reduction = "pca", dims = 1:30, verbose=FALSE)
spike <- FindNeighbors(spike, reduction = "pca", dims = 1:30, verbose=FALSE)
spike <- FindClusters(spike, resolution = 0.3, verbose=FALSE)
# Visualization
p1 <- DimPlot(spike, reduction = "umap", group.by = "seurat_clusters", label = TRUE, repel = TRUE)
p2 <- DimPlot(spike, reduction = "umap", group.by="HTO_classification")
p3 <- DimPlot(spike, reduction="umap", group.by="t-cell")
p4 <- DimPlot(spike, reduction="pca", group.by="patient")

grid.arrange(p1, p2, p3, p4, nrow=2)

# saveRDS(spike, "batch1_combined_spike_seurat.rds")
```
#### Gene expression 
```{r, fig.width=20, fig.height=28}
# spike <- readRDS("batch1_combined_spike_seurat.rds")
VlnPlot(object = spike, features=c("PTPRC", "CCR7", "SELL", "TCF7", "IL7R", "S100A4",
                                  "CD44","CD27", "CD28", "CD69", "CD38","HLA-DRB1",
                                  "IFNG", "CCL4", "CCL5", "CCR4", "CXCR3", "CXCR4",                                                   "SELPLG", "ITGA4", "ITGAL", "ITGB2", "ITGAE", "ICAM1",
                                  "GZMA", "GZMB", "GZMH", "GZMK", "PRF1", "GNLY",
                                  "NKG7", "FASLG", "FAS",  "ISG15", "MX1", "OAS1",
                                  "TUBA1B","LAG3","PDCD1", "HAVCR2", "TIGIT", "CTLA4",
                                 "KLRD1", "FGFBP2", "KLRC1", "KLRC3", "KLRB1", "KLRC2"),
        ncol=6, group.by = "HTO_classification")
```

```{r, fig.width=20, fig.height=24}
VlnPlot(object = spike, features=c("PTPRC", "CCR7", "SELL", "TCF7", "IL7R", "S100A4",
                                  "CD44","CD27", "CD28", "CD69", "CD38","HLA-DRB1",
                                  "IFNG", "CCL4", "CCL5", "CCR4", "CXCR3", "CXCR4",                                                   "SELPLG", "ITGA4", "ITGAL", "ITGB2", "ITGAE", "ICAM1",
                                  "GZMA", "GZMB", "GZMH", "GZMK", "PRF1", "GNLY",
                                  "NKG7", "FASLG", "FAS",  "ISG15", "MX1", "OAS1",
                                  "TUBA1B","LAG3","PDCD1", "HAVCR2", "TIGIT", "CTLA4",
                                 "KLRD1", "FGFBP2", "KLRC1", "KLRC3", "KLRB1", "KLRC2"),
        ncol=6, group.by = "seurat_clusters")
```

```{r, fig.width=12, fig.height=10}
spike <- readRDS("batch1_combined_spike_seurat.rds")

DefaultAssay(spike) <- "RNA"

FeaturePlot(object = spike, features = c("CD4", "CD8A", "CD8B" ), cols = c("grey", "blue"), reduction = "umap")
```

```

