---
title: "SARS-CoV-2 6 month T cells"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(plyr)
library(dplyr)
library(patchwork)
library(cowplot)
library(ggplot2)
library(data.table)
library(tidyverse)
library(harmony)
library(gridExtra)
```

# 10x single cell RNA sequencing from SARS-CoV-2 CD4/8 tetramer sorted T cells

## Batch 2 T1

Dong 050121 T1
Hashtag 1 1131 S151
Hashtag 2 1131 S174
Hashtag 4 1493 S151
Hashtag 5 1493 S174
Hashtag 6 1201 S174
Hashtag 7 1201 S174 - labelled as Hashtag 7 in Excel file but actually Hashtag 8


### Separate Gene Expression and Antibody Capture
```{r}
# Read in data
batch2_t1.counts <- Read10X(data.dir = "/t1-data/user/lfelce/10x_Dong050121/count_T1_repeat2/filtered_feature_bc_matrix")
colnames(batch2_t1.counts) <- gsub(x = colnames(batch2_t1.counts), pattern = "-1", replacement = "")

batch2_t1.hash <- Read10X(data.dir ="/t1-data/user/lfelce/10x_Dong050121/hash_T1_repeat/outs/filtered_feature_bc_matrix")
colnames(batch2_t1.hash) <- gsub(x = colnames(batch2_t1.hash), pattern = "-1", replacement = "")

# Select cell barcodes detected by both RNA and hashtag, then filter
joint.barcodes <- intersect(colnames(batch2_t1.counts), colnames(batch2_t1.hash))

# Subset RNA and hashtag counts by joint cell barcodes
batch2_t1.counts <- batch2_t1.counts[, joint.barcodes]
batch2_t1.hash <- as.matrix(batch2_t1.hash[,joint.barcodes])

# remove Hashtag3 and Hashtag7 (no cells labelled with this Ab)
batch2_t1.hash <- batch2_t1.hash[-c(3,7),]

# rename rows to sample names
rownames(batch2_t1.hash) <- c("1131_CD4_S151", "1131_CD4_S174", 
                              "1493_CD4_S151", "1493_CD4_S174",
                              "1201_CD4_S151", "1201_CD4_S174")

# counts per hashtag
rowSums(batch2_t1.hash)
rowMeans(batch2_t1.hash)

# Set up Seurat object
batch2_t1 <- CreateSeuratObject(counts = batch2_t1.counts)

# Normalize RNA data with log normalization
batch2_t1 <- NormalizeData(batch2_t1)

# Find and scale variable features
batch2_t1 <- FindVariableFeatures(batch2_t1, selection.method = "mean.var.plot")

batch2_t1 <- ScaleData(batch2_t1, features = VariableFeatures(batch2_t1))

# Add HTO data as a new assay independent from RNA
batch2_t1[["HTO"]] <- CreateAssayObject(counts = batch2_t1.hash)

# Normalize HTO data, here we use centered log-ratio (CLR) transformation
batch2_t1 <- NormalizeData(batch2_t1, assay = "HTO", normalization.method = "CLR")

# Demultiplex cells based on hashtag enrichment - use HTODemux() to assign single cells back to sample origins

# If you have a very large dataset we suggest using k_function = 'clara'. This is a k-medoid
# clustering function for large applications You can also play with additional parameters (see
# documentation for HTODemux()) to adjust the threshold for classification Here we are using the
# default settings
batch2_t1 <- HTODemux(batch2_t1, assay = "HTO", positive.quantile = 0.99)
```

```{r, fig.width=12, fig.height=12}
# Global classification results
table(batch2_t1$HTO_classification.global)

# Visualise enrichment for selected hashtags with ridge plots
# Group cells based on max hashtag signal
Idents(batch2_t1) <- "HTO_maxID"
RidgePlot(batch2_t1, assay="HTO", features=rownames(batch2_t1[["HTO"]]), ncol=2)
```

```{r, fig.width=6, fig.height=4}
# Compare number of UMIs for singlets, doublets and negative cells
Idents(batch2_t1) <- "HTO_classification.global"
VlnPlot(batch2_t1, features = "nCount_RNA", pt.size = 0.1, log = TRUE)
```



## Batch 2 T2
Dong 050121 T1
Hashtag 1 1131 NP16
Hashtag 2 1131 ORF3a-28
Hashtag 4 1486 ORF3a-28
Hashtag 5 1502 NP51
Hashtag 6 1201 NP16
Hashtag 8 1004 M24 
All hashtags mixed up?

### Combined Gene Expression and Antibody Capture
```{r}
# Read in data
t2.counts <- Read10X(data.dir = "/t1-data/user/lfelce/10x_Dong050121/combined_T2_repeat/outs/filtered_feature_bc_matrix")

# Remove -1 from barcodes
colnames(t2.counts$`Gene Expression`) <- gsub(x = colnames(t2.counts$`Gene Expression`), pattern = "-1", replacement = "")
colnames(t2.counts$`Antibody Capture`) <- gsub(x = colnames(t2.counts$`Antibody Capture`), pattern = "-1", replacement = "")

# Remove Hashtag 1 and 3 from Antibody Capture (no cells)
# t2.counts$`Antibody Capture` <- t2.counts$`Antibody Capture`[-c(1,3),]

# Rename rows to sample names
# rownames(t2.counts$`Antibody Capture`) <- c("1131_CD8_NP16", 
#                                             "1131_CD8_ORF3a-28", 
#                                             "1502_CD8_NP51",
#                                             "1201_CD8_NP16", "1004_CD4_M24")


# Create Seurat object from Gene Expression
t2 <- CreateSeuratObject(counts = t2.counts$`Gene Expression`)

# Normalize RNA data with log normalization
DefaultAssay(t2) <- "RNA"
t2 <- NormalizeData(t2)

# Find and scale variable features
t2 <- FindVariableFeatures(t2, selection.method = "mean.var.plot")
t2 <- ScaleData(t2, features = VariableFeatures(t2))

# Add HTO data as a new assay independent from RNA
t2[["HTO"]] <- CreateAssayObject(counts = t2.counts$`Antibody Capture`)

# Normalize HTO data, here we use centered log-ratio (CLR) transformation
t2 <- NormalizeData(t2, assay = "HTO", normalization.method = "CLR")

# HTODemux from Seurat too fussy - use MULTIseqDemux, seems to be less fussy about low HTO counts
t2 <- MULTIseqDemux(t2, assay = "HTO", quantile = 0.01)

table(t2$MULTI_ID)

Idents(t2) <- "MULTI_ID"

VlnPlot(t2, features = "nCount_RNA", pt.size = 0.1, log = TRUE)
```


```{r}
# Extract the singlets
t2.singlet <- subset(t2, idents = c("hashtag-1", "hashtag-2", "hashtag-4", "hashtag-5", 
                                    "hashtag-6", "hashtag-7", "hashtag-8"))
# Assign metadata to object
list <- as.data.frame(t2.singlet@meta.data$MULTI_ID)
# colnames(patient_list) <- "sample"
# 
# patient_list$patient <- ifelse(grepl("1131", patient_list$sample), "1131",
#                         ifelse(grepl("1502", patient_list$sample), "1502",
#                         ifelse(grepl("1201", patient_list$sample), "1201",
#                         ifelse(grepl("1004", patient_list$sample), "1004", ""))))
# 
# patient_list$subset <- ifelse(grepl("CD4", patient_list$sample), "CD4",
#                                ifelse(grepl("CD8", patient_list$sample), "CD8", ""))
# 
# patient_list$epitope <- ifelse(grepl("M24", patient_list$sample), "membrane",
#                         ifelse(grepl("ORF3a-28", patient_list$sample), "ORF3a",
#                         ifelse(grepl("NP16", patient_list$sample), "nucleoprotein",
#                         ifelse(grepl("NP51", patient_list$sample), "nucleoprotein", ""))))
# 
# t2.singlet@meta.data$"epitope" <- as.factor(patient_list$epitope)
# t2.singlet@meta.data$"patient" <- as.factor(patient_list$patient)
# t2.singlet@meta.data$"t-cell" <- as.factor(patient_list$subset)

# Select the top 1000 most variable features
t2.singlet <- FindVariableFeatures(t2.singlet, selection.method = "mean.var.plot", verbose=FALSE)

# Scaling RNA data
t2.singlet <- ScaleData(t2.singlet, features = VariableFeatures(t2.singlet), verbose=FALSE)

# Run PCA
t2.singlet <- RunPCA(t2.singlet, features = VariableFeatures(t2.singlet), verbose=FALSE)

# Select top PCs for clustering
t2.singlet <- FindNeighbors(t2.singlet, reduction = "pca", dims = 1:30, verbose=FALSE)
t2.singlet <- FindClusters(t2.singlet, resolution = 0.5, verbose = FALSE)
t2.singlet <- RunUMAP(t2.singlet, dims = 1:30, verbose=FALSE)
t2.singlet <- RunTSNE(t2.singlet, reduction = "pca", dims = 1:30, verbose=FALSE)

# Projecting singlet identities on UMAP visualization
tsne <- DimPlot(t2.singlet, reduction="tsne", group.by = "MULTI_ID")
pca <- DimPlot(t2.singlet, reduction="pca", group.by="MULTI_ID")
umap_cluster <- DimPlot(t2.singlet, reduction="umap", group.by = "seurat_clusters")
# umap_tcell <- DimPlot(t2.singlet, reduction="umap", group.by = "t-cell")
# umap_epitope <- DimPlot(t2.singlet, reduction="umap", group.by = "epitope")
umap_patient <- DimPlot(t2.singlet, reduction="umap", group.by = "MULTI_ID")
```

```{r, fig.width=14, fig.height=8}

grid.arrange(tsne, pca, umap_cluster, umap_patient, nrow=2)

```

```{r, fig.width=10, fig.height=7}
DefaultAssay(t2.singlet) <- "RNA"

FeaturePlot(t2.singlet, features=c("CD8A", "CD8B", "GZMB","CD4"))
```

```{r, fig.width=16, fig.height=10}
DefaultAssay(t2.singlet) <- "HTO"

FeaturePlot(t2.singlet, features=c("hashtag-1", "hashtag-2", "hashtag-4", "hashtag-5", 
                                    "hashtag-6", "hashtag-7", "hashtag-8"))
```

