---
title: "Single-cell RNA-Seq analysis of CD4+ and CD8+ SARS-CoV-2 specific T cells"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(sctransform)
library(SeuratDisk)
library(SeuratData)
library(dplyr)
library(patchwork)
library(cowplot)
library(ggplot2)
library(data.table)
library(tidyverse)
library(plyr)
```

# All samples (except minibulk!)
Analysis of Dong samples: CD8 NP16, CD8 ORF3a-28 and CD4 S34 & M24.
Follow guided clustering tutorial https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html and this one https://scrnaseq-course.cog.sanger.ac.uk/website/seurat-chapter.html


## Load counts matrix and create Seurat object
```{r loading}
setwd('/t1-data/user/lfelce/scRNA-Seq/SmartSeq2_T-cells/')
# load data (from featureCounts)
tcell_data <-  fread('201203_counts.txt', stringsAsFactors = F, header=T)
# remove columns with chromosome, start, end, strand and length info
tcell_data <- tcell_data[,-c(2:6)]
# make Geneid into row names
tcell_data <-tibble::column_to_rownames(tcell_data, "Geneid")
# tidy up sample names
names(tcell_data) <- gsub(x = names(tcell_data), pattern = "./", replacement = "")  
names(tcell_data) <- gsub(x = names(tcell_data), pattern = ".bam", replacement = "")
sample_list <- as.data.frame(colnames(tcell_data))
tcell_data2 <- tcell_data[,-c(481:486,774:779,1404:1409)]

write.table(tcell_data2, file="smartseq_counts.txt", sep="\t")

# Initialise Seurat object with raw (non-normalised) data
# min.cells = include features detected in at least this many cells
# min.features includes cells where at least this many features detected
tcell2 <- CreateSeuratObject(counts = tcell_data2, min.cells = 3, min.features = 200, project = "T-cell_data", assay = "RNA")

```

## Standard pre-processing workflow

### QC and selecting cells for further analysis
Minimum gene cut off already set with CreateSeuratObject, but can filter out cells based on technical or biological parameters.
Visualise gene and molecule counts, plot relationship and exclude cells with clear outlier number of genes detected as potential multiplets (filter user-defined outlier cells). 

Also filter cells based on % mitochondrial genes present:

```{r qc, fig.height=6, fig.width=16}
# non-UMI data - nUMI represents sum of non-normalised values within cells
# calculate % mitochondrial genes and store in separate object
# use object@raw.data represents non-transformed and non-log-normalised counts
# % of UMI mapping to MT-genes is a common scRNA-seq QC metric

mito.genes <- grep(pattern = "^MT-", x = rownames(tcell2@assays[["RNA"]]), value = TRUE)

percent.mito <- Matrix::colSums(tcell2@assays[["RNA"]][mito.genes, ])/Matrix::colSums(tcell2@assays[["RNA"]])

tcell2 <- AddMetaData(object = tcell2, metadata = percent.mito, col.name = "percent.mito") 

VlnPlot(object = tcell2, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), ncol = 3)

# all cells have same value % mitochondrial cells?


```

```{r, fig.height=4, fig.width=6}
# can use GenePlot to visualise gene-gene relationships and anything else calculated by object
FeatureScatter(object=tcell2, feature1="nCount_RNA", feature2="nFeature_RNA")

```


```{r filtering}
# Filter out cells that have unique gene counts (nFeature_RNA) over 4000 or less than 200 
# Note that > and < are used to define a'gate'.  
# -Inf and Inf should be used if don't want a lower or upper threshold.
tcell2 <- subset(x = tcell2, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mito >  -Inf & percent.mito < 0.05 )
```

### Normalising the data
After removing unwanted cells from dataset, need to normalise data. Use global-scaling normalisation method "LogNormalize" - normalises gene expression measurements for each cell by total expression, multiplies by scale factor (10,000 default) and log-tranforms result.
```{r normalisation}
tcell2 <- NormalizeData(object = tcell2, normalization.method = "LogNormalize", scale.factor = 10000)
```

### Detection of variables genes across single cells
Seurat calculates highly variable genes and focuses on these for downstream analysis. FindVariableGenes calculates average expression and dispersion for each gene, places genes into bins and calculates z-score for dispersion within each bin - helps to control relationship between variability and average expression. 

```{r variability, fig.height=8, fig.width = 16}
# identify 2000 variable genes - normal parameter settings for UMI data normalised to total of 1e4 molecules

tcell2 <- FindVariableFeatures(object = tcell2, mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5, nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(tcell2), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(tcell2)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
```
### Scaling data and removing unwanted sources of variation
Dataset may include uninteresting sources of variation - technical noise, batch effects, biological sources of variation (cell cycle) - regressing these signals can improve downstream dimensionality reduction and clustering. To judge effects of these signals, Seurat constructs linear models to predict gene expression based on user-defined variables. Scaled z-scored residuals of these models are stored in scale.data slot, used for dimensionality reduction and clustering.

```{r scaling}
tcell2 <- ScaleData(object = tcell2, vars.to.regress = c("nCounts_RNA", "percent.mito"))

```

### Linear dimensional reduction (PCA)
Perform PCA on scaled data. Not sure if highly variable genes or pc.genes are input?

```{r pca, fig.height=4, fig.width=6}
tcell2 <- RunPCA(object = tcell2,  npcs = 30, verbose = FALSE)
#standard PCA plot
DimPlot(object = tcell2, reduction = "pca")
```

Using a heatmap can be useful to look at primary sources of heterogeneity in a dataset, and can use to decide which PCs to include for further downstream analysis. 

```{r pca heatmap, fig.height=4, fig.width=6}
DimHeatmap(object = tcell2, reduction = "pca", cells = 200, balanced = TRUE)
```

### Determine statistically significant principal components
Overcome extensive technical noise in any single gene for scRNA-seq data, Seurat clusters cells based on PCA scores. Determining how many PCs to include downstream is an important step.

Here it looks like PC1-8 are significant (looking at p-values)
```{r jackstraw, fig.height=4, fig.width=6}
tcell2 <- JackStraw(tcell2, num.replicate = 100)

# visualise distribution of p-values for each PC with uniform distribution (dashed line) Significant PCs show strong enrichment of genes with low p-values (solid curve above dashed line)

tcell2 <- ScoreJackStraw(tcell2, dims = 1:20)

JackStrawPlot(tcell2, dims = 1:15)

```


## Cell clustering
Construct a k-nearest neighbours graph based on Euclidean distance in PCA space, refine edge weights between any 2 cells based on shared overlap in local neighbourhoods. Use Louvain algorithm to cluster cells - modularity optimisation technique to iteratively group cells together.

Resolution parameter - sets granularity of downstream clustering, larger values leads to more clusters. Set between 0.6-1.2 for single cell datasets around 3000 cells. 

```{r clustering}
# calculate KNN and construct SNN graph (find neighbours), then find clusters

tcell2 <- FindNeighbors(tcell2, dims = 1:14)
tcell2 <- FindClusters(tcell2, resolution = 0.2)
```
### Non-linear dimensional reduction (UMAP)
Place similar cells in low-dimensional space. Use same PCs as input to clustering analysis
```{r umap, fig.height=5, fig.width=16}
tcell2 <- RunUMAP(tcell2, dims = 1:14)
# plot UMAP and labels by cluster
Idents(tcell2) <- "seurat_clusters"
UMAP_cluster <- DimPlot(tcell2, reduction = "umap", group.by = "seurat_clusters")

# plot UMAP and label by patient
UMAP_patient <- DimPlot(tcell2, reduction = "umap", group.by = "orig.ident")

UMAP_cluster + UMAP_patient

# save tcell2 object at this point so don't have to redo all of the above again
saveRDS(tcell2, file = "tcell2_seurat.rds")

```

## Finding differentially expressed genes (cluster biomarkers)
Define clusters via differential expression. Identifies positive and negative markers of single cluster compared to all other cells.

```{r number of cells}
# How many cells for each patient?
Idents(tcell2) <- "orig.ident"
table(Idents(tcell2))

# How many cells in each cluster?
Idents(tcell2) <- "seurat_clusters"
table(Idents(tcell2))

```


```{r markers}
# find markers for every cluster compared to remaining cells, report only positive ones
# Wilcoxon Rank Sum test default
tcell2.markers <- FindAllMarkers(object = tcell2, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)

write.csv (tcell2.markers, "sc-tcell_markers.csv")

# divide file into separate clusters

cluster_0 <- tcell2.markers[tcell2.markers$cluster == "0", ]
cluster_1 <- tcell2.markers[tcell2.markers$cluster == "1", ]
cluster_2 <- tcell2.markers[tcell2.markers$cluster == "2", ]
cluster_3 <- tcell2.markers[tcell2.markers$cluster == "3", ]


write.csv (cluster_0, "cluster_0.csv")
write.csv (cluster_1, "cluster_1.csv")
write.csv (cluster_2, "cluster_2.csv")
write.csv (cluster_3, "cluster_3.csv")



# show top 2 differentially expressed genes per cluster
tcell2.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
```

```{r top 2 genes, fig.height=10, fig.width=16}
# Top 2 genes from each cluster - from manual annotation 
VlnPlot(object = tcell2, features =c("CCR7", "CCL5", "IFNG", "BTLA", "PRF1", "GZMH", "OAS1", "IFIT3"), ncol=4)
```


```{r heatmap}
# Expression heatmap for top 20 markers for each cluster
top10 <- tcell2.markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)
# setting slim.col.label to TRUE will print just the cluster IDS instead of every cell name
DoHeatmap(object = tcell2, features = top10$gene, label = TRUE)
```

# Reference mapping - using PBMC 3K dataset
Followed this tutorial https://satijalab.org/seurat/v4.0/reference_mapping.html

Reference dataset - Single Cell Comparison: PBMC 3K data

```{r load pbmc}
# load dataset and process in same way as T cell data
InstallData("pbmc3k")

data("pbmc3k")

pbmc <- pbmc3k

pbmc <- NormalizeData(pbmc)
pbmc <- FindVariableFeatures(pbmc)
pbmc <- ScaleData(pbmc)

```

Find anchors between reference and query using precomputed supervised PCA transformation
```{r anchors}
# find anchors
anchors <- FindTransferAnchors(reference = pbmc, query = tcell2)

# transfer labels
predictions <- TransferData(anchorset = anchors, refdata = pbmc$seurat_annotations)

tcell2 <- AddMetaData (object = tcell2, metadata = predictions)


```
### Run UMAP with cell annotations
```{r patient annotation, fig.height=5, fig.width=8}
# plot UMAP and label by patient
UMAP_patient <- DimPlot(tcell2, reduction = "umap", group.by = "orig.ident") 
UMAP_patient
```

```{r cluster annotation, fig.height=4, fig.width=6}
# plot UMAP and label by cluster
UMAP_cluster <- DimPlot(tcell2, reduction = "umap", group.by = "seurat_clusters") 
UMAP_cluster
```
```{r predicted pbmc id, fig.height=6, fig.width=8}
# plot UMAP and label by annotation
UMAP_annot <- DimPlot(tcell2, reduction = "umap", group.by = "predicted.id", label = TRUE, label.size = 3 ,repel = TRUE)
UMAP_annot
```
```{r cd4 or cd8, fig.height=5, fig.width=8}
# assign known identities to clusters - look at known cell type

cell.type <- rep(c("CD8", "CD4", "CD4", "CD4", "CD8", "CD8", "CD8", "CD8", "CD8", "CD8", "CD4", 
                   "CD4", "CD4", "CD8", "CD8"), 
                 times = c("133", "67", "79", "73", "94", "16", "74","95", "66", "75", "84", 
                           "82", "86", "46", "95"))


tcell2@meta.data$"cell.type" <- as.factor(cell.type)

UMAP_cell <- DimPlot(tcell2, reduction = "umap", group.by = "cell.type", label = TRUE, label.size = 3 ,repel = TRUE)

UMAP_cell

```

```{r severity annotation, fig.height=5, fig.width=8}
# assign known identities to clusters - look at known clinical description

clin.type <- rep(c("mild", "mild", "mild", "severe", "severe", "severe", "severe", "severe", "severe",
                   "mild", "mild", "severe", "severe", "severe", "severe"), 
                 times = c("133", "67", "79", "73", "94", "16", "74","95", "66", "75", "84", 
                           "82", "86", "46", "95"))


tcell2@meta.data$"clin.type" <- as.factor(clin.type)

UMAP_clin <- DimPlot(tcell2, reduction = "umap", group.by = "clin.type", label = TRUE, label.size = 3 ,repel = TRUE)

UMAP_clin

```

```{r known epitopes, fig.height=5, fig.width=8}
# assign known identities to clusters - look at known epitopes

epitope <- rep(c("NP16", "S34", "M24", "S34", "ORF3a-28", "NP16", "NP16", "ORF3a-28", "NP16",
                   "NP16", "M24", "M24", "M24", "ORF3a-28", "ORF3a-28"), 
                 times = c("133", "67", "79", "73", "94", "16", "74","95", "66", "75", "84", 
                           "82", "86", "46", "95"))


tcell2@meta.data$"epitope" <- as.factor(epitope)

UMAP_epitope <- DimPlot(tcell2, reduction = "umap", group.by = "epitope", label = TRUE, label.size = 3 ,repel = TRUE)

# use split.by="seurat_clusters" to also show UMAP by epitope and cluster number

UMAP_epitope

```

```{r timepoints, fig.height=5, fig.width=8}
# assign known identities to clusters - look at known timepoint

timepoint <- rep(c("convalescent", "convalescent", "convalescent", "convalescent", "convalescent",
                   "convalescent", "convalescent", "convalescent", "convalescent", "acute", 
                   "convalescent", "convalescent", "convalescent", "acute", "convalescent"), 
                 times = c("133", "67", "79", "73", "94", "16", "74","95", "66", "75", "84", 
                           "82", "86", "46", "95"))


tcell2@meta.data$"timepoint" <- as.factor(timepoint)

UMAP_timepoint <- DimPlot(tcell2, reduction = "umap", group.by = "timepoint", label = TRUE, label.size = 3 ,repel = TRUE)

UMAP_timepoint

```

## Gene expression by clusters

### T follicular helper
```{r, fig.height=10, fig.width=16}
FeaturePlot(object = tcell2, features = c("IL21", "POU2AF1", "PDCD1", "CXCL13", "CD200", "BTLA", "IL4", "IFNG"), cols = c("grey", "blue"), ncol = 3, reduction = "umap")
```
### Th1
```{r, fig.height=8, fig.width=16}
FeaturePlot(object = tcell2, features = c("TBX21", "IFNG", "LTA", "TNF", "IL2"), cols = c("grey", "blue"), ncol = 3, reduction = "umap")
```

### Th2
```{r, fig.height=4, fig.width=16}
FeaturePlot(object = tcell2, features = c("IL4", "IL5", "IL6", "IL9", "IL13"), cols = c("grey", "blue"), ncol = 3, reduction = "umap")
```
### Treg
```{r, fig.height=4, fig.width=16}
FeaturePlot(object = tcell2, features = c("IL10", "FOXP3", "TGFB1"), cols = c("grey", "blue"), ncol = 3, reduction = "umap")
```
### Memory
```{r, fig.height=12, fig.width=16}
FeaturePlot(object = tcell2, features = c("PTPRC", "CD44", "CD38", "CCR7", "SELL", "TCF7", "IL7R", "S100A4"), cols = c("grey", "blue"), ncol = 3, reduction = "umap")
```
### Activation and differentiation
```{r, fig.height=8, fig.width=16}
FeaturePlot(object = tcell2, features = c("CD27", "CD28", "CD69", "ITGAE", "CD25"), cols = c("grey", "blue"), ncol = 3, reduction = "umap")
```
### Recruitment and migration
```{r, fig.height=8, fig.width=16}
FeaturePlot(object = tcell2, features = c("CCL3", "CCL4", "CCL5", "CXCR3", "CXCR4", "XCR1"), cols = c("grey", "blue"), ncol = 3, reduction = "umap")
```

### Integrins 
```{r, fig.height=10, fig.width=20}
FeaturePlot(object = tcell2, features = c("ITGAE", "ITGB3", "CD107", "ITGB7", "ITGAL", "ITGAD", "ITGB2"), cols = c("grey", "blue"), ncol = 3, reduction = "umap")
```

```{r, fig.height=12, fig.width=18}
FeaturePlot(object = tcell2, features = c("ITGB1","ITGA1", "ITGA4", "ITGA5", "ITGA6", "ICAM3", "ICAM1", "ITGB4", "ALCAM"), cols = c("grey", "blue"), ncol = 3, reduction = "umap")
```

### Type I interferon
```{r, fig.height=8, fig.width=18}
FeaturePlot(object = tcell2, features = c("OAS1", "MX1", "IFIT3", "IFI44L", "ISG15", "MX2"), cols = c("grey", "blue"), ncol = 3, reduction = "umap")
```
### Cytotoxicity
```{r, fig.height=10, fig.width=18}
FeaturePlot(object = tcell2, 
            features = c("GZMA", "GZMB", "GZMH", "GZMK", "GZMM", "PRF1", "CSF2", "GNLY"), 
            cols = c("grey", "blue"), reduction = "umap", ncol= 3)
```


## TCR information

```{r load tcr gene info}
cd4_tcr <- read.csv("cd4_tcr.csv", row.names=1)
cd8_np16_tcr <- read.csv("cd8_np16_tcr.csv", row.names=1)
cd8_orf_tcr <- read.csv("cd8_orf_tcr.csv", row.names=1)

tcr <- rbind(cd4_tcr, cd8_np16_tcr, cd8_orf_tcr)

tcr <- tcr[order(tcr$cell_name),]
tcr <- tcr[,c(1,3,4,6,7)]

new_names <- as.data.frame(rownames(tcell2@meta.data))
colnames(new_names)[1] <- "cell_name"
new_names <- tibble::rownames_to_column(new_names, "order")

tcr_edit <-  join(new_names, tcr)

trav <- tcr_edit[3]
traj <- tcr_edit[4]
trbv <- tcr_edit[5]
trbj <- tcr_edit[6]

tcell2@meta.data$"trav" <- as.factor(trav$TRAV)
tcell2@meta.data$"traj" <- as.factor(traj$TRAJ)
tcell2@meta.data$"trbv" <- as.factor(trbv$TRBV)
tcell2@meta.data$"trbj" <- as.factor(trbj$TRBJ)

```

```{r trav umap, fig.height=6, fig.width=10}
UMAP_trav <- DimPlot(tcell2, reduction = "umap", group.by = "trav", label = TRUE, label.size = 3 ,repel = TRUE)

UMAP_trav

```

```{r traj umap, fig.height=6, fig.width=10}
UMAP_traj <- DimPlot(tcell2, reduction = "umap", group.by = "traj", label = TRUE, label.size = 3 ,repel = TRUE)

UMAP_traj

```
```{r trbv umap, fig.height=6, fig.width=10}
UMAP_trbv <- DimPlot(tcell2, reduction = "umap", group.by = "trbv", label = TRUE, label.size = 3 ,repel = TRUE)+ NoLegend()

UMAP_trbv

```
```{r trbj umap, fig.height=6, fig.width=10}
UMAP_trbj <- DimPlot(tcell2, reduction = "umap", group.by = "trbj", label = TRUE, label.size = 3 ,repel = TRUE)

UMAP_trbj

```

```{r investigating cluster 3, fig.height=4, fig.width=16}
# compare specifically to 0 and 2 (neighbouring clusters)
cluster3.markers <- FindMarkers(tcell2, ident.1 = 3, ident.2 = c(0,2), min.pct = 0.25)

head(cluster3.markers, n = 10)

FeaturePlot(object = tcell2, features = c("FOXP3", "MX1"), cols = c("grey", "blue", "black"), ncol = 2, reduction = "umap", blend = TRUE)

# FOXP3 not on top list when compared with neighbouring clusters, only IL2RA, IFIT3 and OAS1. When looking at blended UMAP, can see that FOXP3 only expressed in 2 cells in cluster 3, and not same cells as MX1. Therefore probably not that many Tregs.

```


# Save file

```{r save}
# save tcell2 object at this point so don't have to redo all of the above again
saveRDS(tcell2, file = "tcell2_seurat.rds")

```
