---
title: "TCR Repertoire Diversity"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tcR)
library(data.table)
library(stringr)
library(plyr)
library(dplyr)
library(cowplot)
library(tidyverse)
library(gridExtra)
library(ggplot2)

```

TCR Repertoire Diversity looking at CDR3 sequences for alpha and beta chains separately, using MiXCR output from SmartSeq2 scRNA-Seq, and calculating diversity using Shannon diversity index (tcR package)
https://cran.r-project.org/web/packages/tcR/vignettes/tcrvignette.html

repDiversity from tcR package not working properly - try using baseR to calculate Shannon Diversity Index https://en.wikipedia.org/wiki/Diversity_index

# CD8 NP16 and ORF3a-28 (combined) - tcR
Only interested in 1 beta (all cells)
```{r, text=FALSE, include=FALSE}
# # load in data and create mixcr_a and mixcr_b object
# setwd('/t1-data/user/lfelce/TCR_analysis/cd8_np16_new/')
# 
# # create list of file names
# # tra_list <- list.files(path = ".", recursive = TRUE,
# #                        pattern = "\\TRA.txt$", 
# #                        full.names = TRUE)
# # tra_list <- tra_list %>% str_replace("./*", "")
# 
# trb_list <- list.files(path = ".", recursive = TRUE,
#                        pattern = "\\TRB.txt$", 
#                        full.names = TRUE)
# trb_list <- trb_list %>% str_replace("./*", "")
# 
# 
# # parse in files with tcR
# # mixcr_a_np16 <- parse.file.list(tra_list, "mixcr")
# mixcr_b_np16 <- parse.file.list(trb_list, "mixcr")
# 
# # load in data and create mixcr_a and mixcr_b object
# setwd('/t1-data/user/lfelce/TCR_analysis/cd8_orf_new/')
# 
# # create list of file names, remove minibulk
# # tra_list <- list.files(path = ".", recursive = TRUE,
# #                        pattern = "\\TRA.txt$", 
# #                        full.names = TRUE)
# # tra_list <- tra_list %>% str_replace("./*", "")
# # tra_list <-tra_list[!str_detect(tra_list,pattern="minibulk")]
# 
# trb_list <- list.files(path = ".", recursive = TRUE,
#                        pattern = "\\TRB.txt$", 
#                        full.names = TRUE)
# trb_list <- trb_list %>% str_replace("./*", "")
# trb_list <-trb_list[!str_detect(trb_list,pattern="minibulk")]
# 
# 
# # parse in files with tcR
# # mixcr_a_orf3a <- parse.file.list(tra_list, "mixcr")
# mixcr_b_orf3a <- parse.file.list(trb_list, "mixcr")
# 
# # combine mixcr files
# # mixcr_a <- c(mixcr_a_np16, mixcr_a_orf3a)
# mixcr_b <- c(mixcr_b_np16, mixcr_b_orf3a)
# 
# # sort alphabetically
# # mixcr_a <- mixcr_a[order(names(mixcr_a))]
# mixcr_b <- mixcr_b[order(names(mixcr_b))]

# # create list of cell numbers and cell names
# # different lengths so same cell will be different number in a or b
# # mixcr_a_names <- as.data.frame(names(mixcr_a))
# mixcr_b_names <- as.data.frame(names(mixcr_b))
# # mixcr_a_names <-tibble::rownames_to_column(mixcr_a_names, "cell_number")
# mixcr_b_names <- tibble::rownames_to_column(mixcr_b_names, "cell_number")
# 
# # rename columns
# # colnames(mixcr_a_names) <- c("cell_number", "cell_name")
# colnames(mixcr_b_names) <- c("cell_number", "cell_name")

# mixcr_a
# datalist = list()
# for (i in (1:length(mixcr_a))) {
#   dat <- data.frame(c(mixcr_a[[i]][3], mixcr_a[[i]][4], mixcr_a[[i]][6],mixcr_a[[i]][7], mixcr_a[[i]][8]))
#   dat$i <- i # keep track of which iteration produced it
#   datalist[[i]] <- dat # add it to list
# }
# combine columns for each cell and filter to keep cells with 1 or 2 alphas
# big_data = do.call(rbind, datalist)
# tra <- big_data %>% group_by(i) %>% filter(n() <= 2)
# colnames(tra) <- c("clone_count", "clone_fraction", "CDR3_alpha", "TRAV", "TRAJ", "cell_number")
# tra <- merge(tra, mixcr_a_names, by="cell_number")


# # mixcr_b
# datalist = list()
# for (i in (1:length(mixcr_b))) {
#   dat <- data.frame(c(mixcr_b[[i]][3], mixcr_b[[i]][4], mixcr_b[[i]][6], mixcr_b[[i]][7], mixcr_b[[i]][8]))
#   dat$i <- i # keep track of which iteration produced it
#   datalist[[i]] <- dat # add it to list
# }
# # combine columns for each cell
# # have to have 1 or 2 betas per cell - if only keep cells with 1 beta then diversity index = 0 for all betas!
# big_data = do.call(rbind, datalist)
# trb <- big_data %>% group_by(i) %>% filter(n() <= 2)
# 
# colnames(trb) <- c("clone_count", "clone_fraction","CDR3_beta", "TRBV", "TRBJ", "cell_number")
# trb <- merge(trb, mixcr_b_names, by="cell_number")
# 
# # tidy up
# # tra <- mutate(tra, alpha=paste(TRAV, TRAJ, sep="_"))
# 
# trb <- mutate(trb, beta=paste(TRBV, TRBJ, sep="_"))
```

### Shannon diversity box plots for CDR3 alpha and beta
```{r, text=FALSE}
## alpha
# calculate diversity
# div <- as.data.frame(repDiversity(.data = mixcr_a, .method = "entropy", .quant = "read.count"))
# 
# div_a <- tibble::rownames_to_column(div, "cell_name")
# colnames(div_a ) <- c("cell_name", "diversity_index")
# 
# # add patient info and timepoint
# div_a$patient <- ifelse(grepl("005", div_a$cell_name), "005", 
#                           ifelse(grepl("1131-TP-1",div_a$cell_name), "1131-TP-1", 
#                           ifelse(grepl("1131-TP-2", div_a$cell_name), "1131-TP-2", 
#                           ifelse(grepl("1153", div_a$cell_name), "1153",
#                           ifelse(grepl("1201", div_a$cell_name), "1201",
#                           ifelse(grepl("1105", div_a$cell_name), "1105",
#                           ifelse(grepl("1134-TP-2",div_a$cell_name), "1134-TP-2",   
#                           ifelse(grepl("1525-TP-1", div_a$cell_name), "1525-TP-1",
#                           ifelse(grepl("1525-TP-2", div_a$cell_name), "1525-TP-2", 
#                                  "")))))))))
# 
# div_a$severity <- ifelse(grepl("005", div_a$cell_name), "mild", 
#                           ifelse(grepl("1131-TP-1",div_a$cell_name), "severe", 
#                           ifelse(grepl("1131-TP-2", div_a$cell_name), "severe", 
#                           ifelse(grepl("1153", div_a$cell_name), "severe",
#                           ifelse(grepl("1201", div_a$cell_name), "mild",
#                           ifelse(grepl("1105", div_a$cell_name), "severe",
#                           ifelse(grepl("1134-TP-2",div_a$cell_name), "severe",   
#                           ifelse(grepl("1525-TP-1", div_a$cell_name), "severe",
#                           ifelse(grepl("1525-TP-2", div_a$cell_name), "severe", 
#                                  "")))))))))
# 
# div_a$timepoint <- ifelse(grepl("TP-1", div_a$patient), "acute",
#                    ifelse(grepl("005|TP-2|1153|1201|1105", div_a$cell_name), "convalescent", ""))
# 
# div_a$epitope <- ifelse(grepl("NP16|B7_SPR", div_a$cell_name), "NP16", 
#                  ifelse(grepl("ORF3a-28", div_a$cell_name), "ORF3a-28", ""))
# 
# # merge with TCR info
# div_a <- merge(div_a, tra, by="cell_name")
# 
# # view div_a, select only acute samples, copy to Excel
# # remove duplicated rows and then check in R view(div_a) which are shared or acute only
# 
# # add TCR status
# div_a$TCR <- ifelse(grepl("TRAV21_TRAJ37|TRAV17_TRAJ22|TRAV13-1_TRAJ11|TRAV21_TRAJ30|TRAV13-1_TRAJ4|TRAV8-6_TRAJ22|TRAV14DV4_TRAJ9|TRAV17_TRAJ7|TRAV35_TRAJ52|TRAV20_TRAJ35|TRAV14DV4_TRAJ43|TRAV14DV4_TRAJ52|TRAV24_TRAJ22|TRAV25_TRAJ47", div_a$alpha), "shared",
#              ifelse(grepl("TRAV13-2_TRAJ12|TRAV26-1_TRAJ4|TRAV26-1_TRAJ40|TRAV26-1_TRAJ28|TRAV12-1_TRAJ38|TRAV14DV4_TRAJ8|TRAV12-1|TRAJ18|TRAV12-3_TRAJ36|TRAV8-4, TRAV8-6_TRAJ22|TRAV25_TRAJ24|TRAV25_TRAJ15|TRAV12-2_TRAJ9", div_a$alpha), "acute only",
#              ifelse(!grepl("TRAV21_TRAJ37|TRAV17_TRAJ22|TRAV13-1_TRAJ11|TRAV21_TRAJ30|TRAV13-1_TRAJ4|TRAV8-6_TRAJ22|TRAV14DV4_TRAJ9|TRAV17_TRAJ7|TRAV35_TRAJ52|TRAV20_TRAJ35|TRAV14DV4_TRAJ43|TRAV14DV4_TRAJ52|TRAV24_TRAJ22|TRAV25_TRAJ47|RAV13-2_TRAJ12|TRAV26-1_TRAJ4|TRAV26-1_TRAJ40|TRAV26-1_TRAJ28|TRAV12-1_TRAJ38|TRAV14DV4_TRAJ8|TRAV12-1|TRAJ18|TRAV12-3_TRAJ36|TRAV8-4, TRAV8-6_TRAJ22|TRAV25_TRAJ24|TRAV25_TRAJ15|TRAV12-2_TRAJ9", div_a$alpha), "convalescent only", "")))



# # beta
# div <- as.data.frame(repDiversity(.data = mixcr_b, .method = "entropy", .quant = "read.count"))
# 
# div_b <- tibble::rownames_to_column(div, "cell_name")
# colnames(div_b ) <- c("cell_name", "diversity_index")
# 
# # add patient info and timepoint
# div_b$patient <- ifelse(grepl("005", div_b$cell_name), "005", 
#                           ifelse(grepl("1131-TP-1",div_b$cell_name), "1131-TP-1", 
#                           ifelse(grepl("1131-TP-2", div_b$cell_name), "1131-TP-2", 
#                           ifelse(grepl("1153", div_b$cell_name), "1153",
#                           ifelse(grepl("1201", div_b$cell_name), "1201",
#                           ifelse(grepl("1105", div_b$cell_name), "1105",
#                           ifelse(grepl("1134-TP-2",div_b$cell_name), "1134-TP-2",   
#                           ifelse(grepl("1525-TP-1", div_b$cell_name), "1525-TP-1",
#                           ifelse(grepl("1525-TP-2", div_b$cell_name), "1525-TP-2", 
#                                  "")))))))))
# 
# div_b$severity <- ifelse(grepl("005", div_b$cell_name), "mild", 
#                           ifelse(grepl("1131-TP-1",div_b$cell_name), "severe", 
#                           ifelse(grepl("1131-TP-2", div_b$cell_name), "severe", 
#                           ifelse(grepl("1153", div_b$cell_name), "severe",
#                           ifelse(grepl("1201", div_b$cell_name), "mild",
#                           ifelse(grepl("1105", div_b$cell_name), "severe",
#                           ifelse(grepl("1134-TP-2",div_b$cell_name), "severe",   
#                           ifelse(grepl("1525-TP-1", div_b$cell_name), "severe",
#                           ifelse(grepl("1525-TP-2", div_b$cell_name), "severe", 
#                                  "")))))))))
# 
# div_b$timepoint <- ifelse(grepl("TP-1", div_b$patient), "acute",
#                    ifelse(grepl("005|TP-2|1153|1201|1105", div_b$cell_name), "convalescent", ""))
# 
# div_b$epitope <- ifelse(grepl("NP16|B7_SPR", div_b$cell_name), "NP16", 
#                  ifelse(grepl("ORF3a-28", div_b$cell_name), "ORF3a-28", ""))
# 
# 
# # merge with TCR info
# div_b <- merge(div_b, trb, by="cell_name")
# 
# # remove 2nd row of cells with 2 betas - actually checked cells with 2 betas, CDR3b sequences are exactly the same (and therefore have same diversity index), but have different gene usage so will decrease overall numbers but shouldn't affect actual diversity?
# # 504 -> 380 unique cells
# div_b <- div_b[!duplicated(div_b$cell_name),]
# 
# 
# # add TCR status
# div_b$TCR <- ifelse(grepl("TRBV27_TRBJ2-4|TRBV28_TRBJ1-1|TRBV5-1_TRBJ2-1|TRBV13_TRBJ1-2|TRBV4-2_TRBJ2-1|TRBV5-1_TRBJ2-5|TRBV4-1_TRBJ2-3|TRBV6-6_TRBJ2-7|TRBV4-1_TRBJ2-7|TRBV5-1_TRBJ2-7|TRBV2_TRBJ2-2|TRBV4-1, TRBV4-3_TRBJ2-7", div_b$beta), "shared",
#              ifelse(grepl("TRBV29-1_TRBJ2-1|TRBV9_TRBJ2-3|TRBV7-9_TRBJ1-6|TRBV28_TRBJ2-7|TRBV7-1, TRBV28_TRBJ2-7|TRBV6-1, TRBV6-5, TRBV6-6_TRBJ2-6", div_b$beta), "acute only",
#              ifelse(!grepl("TRBV27_TRBJ2-4|TRBV28_TRBJ1-1|TRBV5-1_TRBJ2-1|TRBV13_TRBJ1-2|TRBV4-2_TRBJ2-1|TRBV5-1_TRBJ2-5|TRBV4-1_TRBJ2-3|TRBV6-6_TRBJ2-7|TRBV4-1_TRBJ2-7|TRBV5-1_TRBJ2-7|TRBV2_TRBJ2-2|TRBV4-1, TRBV4-3_TRBJ2-7|TRBV29-1_TRBJ2-1|TRBV9_TRBJ2-3|TRBV7-9_TRBJ1-6|TRBV28_TRBJ2-7|TRBV7-1, TRBV28_TRBJ2-7|TRBV6-1, TRBV6-5, TRBV6-6_TRBJ2-6", div_b$beta), "convalescent only", "")))



# export files
# setwd('/t1-data/user/lfelce/TCR_analysis/diversity/')
# write.csv(div_a, "cd8_np16_orf3a-28_div_a.csv")
# write.csv(div_b, "cd8_np16_orf3a-28_div_b.csv")
```

```{r}
# groups <- read.csv("/t1-data/user/lfelce/scRNA-Seq/SmartSeq2_T-cells/cd8_np16_orf3a_sc_groups.csv", header=TRUE)
# 
# # add 0 in front of single digit numbers
# groups$new.group <- str_pad(groups$new.group, 2, side="left", pad="0")
# 
# div_a_groups <- merge(div_a, groups, by.x="cell_name", by.y="name", all.x=TRUE)
# 
# div_a_groups <- mutate(div_a_groups, group=paste(epitope, new.group, sep="_"))
```


```{r}
# p1 <- ggplot(div_b, aes(x=patient, y=diversity_index, color=TCR)) + 
#   geom_jitter(width=0.25) +
#   labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Patient", y = "Shannon diversity index") 
#   
# p2 <- ggplot(div_b, aes(x=timepoint, y=diversity_index, color=TCR)) + 
#   geom_jitter(width=0.25)+
#   labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Timepoint", y = "Shannon diversity index")
# 
# p3 <- ggplot(div_b, aes(x=epitope, y=diversity_index, color=TCR)) + 
#   geom_jitter(width=0.25)+
#   labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Epitope", y = "Shannon diversity index")


```

#### By patient
```{r, fig.width=10, fig.height=5}
# p1
```
#### By timepoint and epitope

```{r, fig.width=16, fig.height=7}
# grid.arrange(p2, p3, nrow=1)
```
#### By severity
Can't really compare mild v severe as both mild cases in NP16?
```{r, fig.height=6, fig.width=18}
# p7 <- ggplot(div_a, aes(x=severity, y=diversity_index,color=TCR)) + 
#   geom_jitter(width=0.25)+
#   labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 alpha",x="Severity", y = "Shannon diversity index")
# 
# p8 <- ggplot(div_b, aes(x=severity, y=diversity_index,color=TCR)) + 
#   geom_jitter(width=0.25)+
#   labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Severity", y = "Shannon diversity index")
# 
# grid.arrange(p7, p8, nrow=1)
```



```{r, fig.width=12, fig.height=7}
# # Welch 2 sample t-test: (y~x) where y is numeric and x is a binary factor
# 
# t.test(div_b$diversity_index~div_b$epitope)
# t.test(div_b$diversity_index~div_b$timepoint)
# 
# q1 <- ggplot(div_b, aes(x=epitope, y=diversity_index)) +
#   geom_boxplot()+
# labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Epitope", y = "Shannon diversity index") 
# q2 <- ggplot(div_b, aes(x=timepoint, y=diversity_index)) +
#   geom_boxplot()+
# labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Timepoint", y = "Shannon diversity index")
# 
# grid.arrange(q1, q2, nrow=1)

```

```{r}
# ggplot(div_a_groups, aes(x=severity, y=diversity_index, color=group)) + 
#   geom_jitter(width=0.25) +
#   labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 alpha",x="Patient", y = "Shannon diversity index") 
```


# NP16 and ORF3a-28 combined only from 1/2 alphas + 1 beta cells (complete single cell data) - baseR

```{r, text=FALSE, include=FALSE}
# # NP16 data from single cell TCRs - 1 beta cells only (1/2 alphas)
# sc_np16 <- read.csv("/t1-data/user/lfelce/TCR_analysis/new_mixcr_results/cd8_np16_complete_sc_tcr.csv")
# 
# sc_np16_b <- sc_np16[,c("cell_name", "CDR3_beta", "beta")]
# 
# sc_np16_b <- sc_np16_b[!duplicated(sc_np16_b$cell_name),]
# 
# 
# 
# # ORF3a-28 data from single cell TCRs - 1 beta cells only (1/2 alphas)
# sc_orf3a <- read.csv("/t1-data/user/lfelce/TCR_analysis/new_mixcr_results/cd8_orf3a-28_complete_sc_tcr.csv")
# 
# sc_orf3a_b <- sc_orf3a[,c("cell_name", "CDR3_beta", "beta")]
# 
# sc_orf3a_b <- sc_orf3a_b[!duplicated(sc_orf3a_b$cell_name),]


```

## Shannon diversity index (base R)
Email from Benny:

Diversity usually applies to a repertoire, not an individual TCR. It refers to the distribution of frequencies – how MANY different TCRs are there, and how are they distributed. For example are they all at equal number, or are there a few very frequent and others very rare. It is not usually used to measure the sort of diversity you mean, which is how similar are the TCRs. Have a look at this paper, which may be helpful. https://pubmed.ncbi.nlm.nih.gov/32952933/

 There are many different measures of the first type of diversity I mentioned above, depending on how much weight you want to give to the first question (how MANY) or the second (how are they distributed). For example, RICHNESS is simply a measure of how many different TCRs there are as a proportion fo the total number. The other extreme is to measure what proportion of the reprtorie is occupied by the biggest clone.

Two commonly used intermediates between these two are Shannon and Simpson. You don’t really need a package for Shannon – it is simply sum(pi*log(p)) taken over all the TCRs, where p is the proportion of the repertoire occupied by clone i. But you can use vegan v2.4-2by Jari Oksanen which will give you Shannon and Simpson.

NOTE – diversity indices are very dependent on total reprtorie size. The only real way to deal with this is to subsample each reprotire you are comparing to the same number of TCRs. You can try a normalised Shannon (which divides the total by log(N) but its still not perfect.

```{r}

# combined <- rbind(sc_np16_b, sc_orf3a_b)
# 
# table <- as.data.frame(table(combined$CDR3_beta))
# 
# combined2 <- merge(combined, table, by.x="CDR3_beta", by.y="Var1", all.x=TRUE)
# 
# combined2$patient <- ifelse(grepl("005", combined2$cell_name), "005",
#                           ifelse(grepl("1131-TP-1",combined2$cell_name), "1131-TP-1",
#                           ifelse(grepl("1131-TP-2", combined2$cell_name), "1131-TP-2",
#                           ifelse(grepl("1153", combined2$cell_name), "1153",
#                           ifelse(grepl("1201", combined2$cell_name), "1201",
#                           ifelse(grepl("1105", combined2$cell_name), "1105",
#                           ifelse(grepl("1134-TP-2",combined2$cell_name), "1134-TP-2",
#                           ifelse(grepl("1525-TP-1", combined2$cell_name), "1525-TP-1",
#                           ifelse(grepl("1525-TP-2", combined2$cell_name), "1525-TP-2",
#                                  "")))))))))
# 
# combined2$severity <- ifelse(grepl("005", combined2$cell_name), "mild",
#                           ifelse(grepl("1131-TP-1",combined2$cell_name), "severe",
#                           ifelse(grepl("1131-TP-2", combined2$cell_name), "severe",
#                           ifelse(grepl("1153", combined2$cell_name), "severe",
#                           ifelse(grepl("1201", combined2$cell_name), "mild",
#                           ifelse(grepl("1105", combined2$cell_name), "severe",
#                           ifelse(grepl("1134-TP-2",combined2$cell_name), "severe",
#                           ifelse(grepl("1525-TP-1", combined2$cell_name), "severe",
#                           ifelse(grepl("1525-TP-2", combined2$cell_name), "severe",
#                                  "")))))))))
# 
# combined2$timepoint <- ifelse(grepl("TP-1", combined2$patient), "acute",
#                    ifelse(grepl("005|TP-2|1153|1201|1105", combined2$cell_name), "convalescent", ""))
# 
# combined2$epitope <- ifelse(grepl("NP16|B7_SPR", combined2$cell_name), "NP16",
#                  ifelse(grepl("ORF3a-28", combined2$cell_name), "ORF3a-28", ""))
# 
# datalist = list()
# for (i in (1:nrow(combined2))) {
#   dat <- combined2$Freq[i]/sum(combined2$Freq)
#   datalist[[i]] <- dat # add it to list
# }
# 
# proportion <- do.call(rbind, datalist)
# 
# combined2$proportion <- proportion
# 
# datalist = list()
# for (i in (1:nrow(combined2))) {
#   dat <- -sum(combined2$proportion[i]*(log10(combined2$proportion[i])))
#   datalist[[i]] <- dat # add it to list
# }
# 
# diversity <- do.call(rbind, datalist)
# 
# combined2$diversity <- diversity

                           
```

```{r}
# p1 <- ggplot(combined2, aes(x=patient, y=diversity, colour=epitope)) +
#   geom_boxplot(width=0.5) + geom_jitter(width=0.25) +
#   labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Patient", y = "Shannon diversity index")
# 
# p2 <- ggplot(combined2, aes(x=epitope, y=diversity, colour=epitope)) +
#   geom_boxplot(width=0.5)+ geom_jitter(width=0.25) +
#   labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Epitope", y = "Shannon diversity index")
# 
# # subset NP16 only to look at mild v severe
# 
# np16_only <- combined2[!grepl("ORF3a-28", combined2$epitope),]
# 
# np16_only$proportion <- NULL
# np16_only$diversity <- NULL
# 
# datalist = list()
# for (i in (1:nrow(np16_only))) {
#   dat <- np16_only$Freq[i]/sum(np16_only$Freq)
#   datalist[[i]] <- dat # add it to list
# }
# 
# proportion <- do.call(rbind, datalist)
# 
# np16_only$proportion <- proportion
# 
# datalist = list()
# for (i in (1:nrow(np16_only))) {
#   dat <- -sum(np16_only$proportion[i]*(log10(np16_only$proportion[i])))
#   datalist[[i]] <- dat # add it to list
# }
# 
# diversity <- do.call(rbind, datalist)
# 
# np16_only$diversity <- diversity
# 
# p3 <- ggplot(np16_only, aes(x=severity, y=diversity, colour=severity)) +
#   geom_boxplot(width=0.5)+ geom_jitter(width=0.25) +
#   labs(title="Diversity of CD8+ NP16 CDR3 beta",x="Severity", y = "Shannon diversity index")


```


#### By patient
```{r, fig.width=10, fig.height=5}
# p1
```
#### By severity and epitope

```{r, fig.width=16, fig.height=7}
# grid.arrange(p2, p3, nrow=1)
```



```{r}
# # Welch 2 sample t-test: (y~x) where y is numeric and x is a binary factor
# 
# t.test(combined2$diversity~combined2$epitope)
# 
# t.test(np16_only$diversity~np16_only$severity)


```

# NP16 and ORF3a-28 combined cells with 1 beta only (ignoring alpha)
TCr diversity looking at each repertoire separately (NP16 and ORF3a-28) to see which repertoire has more unique sequences and how often they occur - looking at each CDR3 beta sequence to see what proportion they occur (frequency/total number of cells) CDR3 beta sequence with higher proportion = higher diversity index

```{r, include=FALSE}
# # load in data and create mixcr_b object
# setwd('/t1-data/user/lfelce/TCR_analysis/cd8_np16_new/')
# 
# # create list of file names
# trb_list <- list.files(path = ".", recursive = TRUE,
#                         pattern = "\\TRB.txt$",
#                         full.names = TRUE)
# trb_list <- trb_list %>% str_replace("./*", "")
# 
# # parse in files with tcR
# mixcr_b_np16 <- parse.file.list(trb_list, "mixcr")
# 
# # load in data and create  mixcr_b object
# setwd('/t1-data/user/lfelce/TCR_analysis/cd8_orf_new/')
# 
# # create list of file names, remove minibulk
# trb_list <- list.files(path = ".", recursive = TRUE,
#                        pattern = "\\TRB.txt$",
#                        full.names = TRUE)
# trb_list <- trb_list %>% str_replace("./*", "")
# trb_list <-trb_list[!str_detect(trb_list,pattern="minibulk")]
# 
# # parse in files with tcR
# mixcr_b_orf3a <- parse.file.list(trb_list, "mixcr")
# 
# # combine mixcr files
# mixcr_b <- c(mixcr_b_np16, mixcr_b_orf3a)
# 
# # sort alphabetically
# mixcr_b <- mixcr_b[order(names(mixcr_b))]
# 
# # create list of cell numbers and cell names
# # different lengths so same cell will be different number in a or b
# mixcr_b_names <- as.data.frame(names(mixcr_b))
# mixcr_b_names <- tibble::rownames_to_column(mixcr_b_names, "cell_number")
# 
# # rename columns
# colnames(mixcr_b_names) <- c("cell_number", "cell_name")
# 
# # mixcr_b
# datalist = list()
# for (i in (1:length(mixcr_b))) {
#   dat <- data.frame(c(mixcr_b[[i]][3], mixcr_b[[i]][4], mixcr_b[[i]][6], mixcr_b[[i]][7], mixcr_b[[i]][8]))
#   dat$i <- i # keep track of which iteration produced it
#   datalist[[i]] <- dat # add it to list
# }
# # combine columns for each cell
# # 1 beta cell only
# big_data = do.call(rbind, datalist)
# trb <- big_data %>% group_by(i) %>% filter(n() == 1)
# 
# colnames(trb) <- c("clone_count", "clone_fraction","CDR3_beta", "TRBV", "TRBJ", "cell_number")
# trb <- merge(trb, mixcr_b_names, by="cell_number")
# 
# # tidy up
# trb <- mutate(trb, beta=paste(TRBV, TRBJ, sep="_"))
# 
# table <- as.data.frame(table(trb$CDR3_beta))
# 
# combined2 <- merge(trb, table, by.x="CDR3_beta", by.y="Var1", all.x=TRUE)
# 
# combined2$patient <- ifelse(grepl("005", combined2$cell_name), "005",
#                           ifelse(grepl("1131-TP-1",combined2$cell_name), "1131-TP-1",
#                           ifelse(grepl("1131-TP-2", combined2$cell_name), "1131-TP-2",
#                           ifelse(grepl("1153", combined2$cell_name), "1153",
#                           ifelse(grepl("1201", combined2$cell_name), "1201",
#                           ifelse(grepl("1105", combined2$cell_name), "1105",
#                           ifelse(grepl("1134-TP-2",combined2$cell_name), "1134-TP-2",
#                           ifelse(grepl("1525-TP-1", combined2$cell_name), "1525-TP-1",
#                           ifelse(grepl("1525-TP-2", combined2$cell_name), "1525-TP-2",
#                                  "")))))))))
# 
# combined2$severity <- ifelse(grepl("005", combined2$cell_name), "mild",
#                           ifelse(grepl("1131-TP-1",combined2$cell_name), "severe",
#                           ifelse(grepl("1131-TP-2", combined2$cell_name), "severe",
#                           ifelse(grepl("1153", combined2$cell_name), "severe",
#                           ifelse(grepl("1201", combined2$cell_name), "mild",
#                           ifelse(grepl("1105", combined2$cell_name), "severe",
#                           ifelse(grepl("1134-TP-2",combined2$cell_name), "severe",
#                           ifelse(grepl("1525-TP-1", combined2$cell_name), "severe",
#                           ifelse(grepl("1525-TP-2", combined2$cell_name), "severe",
#                                  "")))))))))
# 
# combined2$timepoint <- ifelse(grepl("TP-1", combined2$patient), "acute",
#                    ifelse(grepl("005|TP-2|1153|1201|1105", combined2$cell_name), "convalescent", ""))
# 
# combined2$epitope <- ifelse(grepl("NP16|B7_SPR", combined2$cell_name), "NP16",
#                  ifelse(grepl("ORF3a-28", combined2$cell_name), "ORF3a-28", ""))
# 
# 
# # subset NP16 only to look at mild v severe
# 
# np16 <- combined2[!grepl("ORF3a-28", combined2$epitope),]
# 
# np16$proportion <- NULL
# np16$diversity <- NULL
# 
# datalist = list()
# for (i in (1:nrow(np16))) {
#   dat <- np16$Freq[i]/nrow(np16)
#   datalist[[i]] <- dat # add it to list
# }
# 
# proportion <- do.call(rbind, datalist)
# 
# np16$proportion <- proportion
# 
# datalist = list()
# for (i in (1:nrow(np16))) {
#   dat <- -sum(np16$proportion[i]*(log10(np16$proportion[i])))
#   datalist[[i]] <- dat # add it to list
# }
# 
# diversity <- do.call(rbind, datalist)
# 
# np16$diversity <- diversity
# 
# # subset ORF3a-28 only to look at mild v severe
# 
# orf3a <- combined2[!grepl("NP16", combined2$epitope),]
# 
# orf3a$proportion <- NULL
# orf3a$diversity <- NULL
# 
# datalist = list()
# for (i in (1:nrow(orf3a))) {
#   dat <- orf3a$Freq[i]/nrow(orf3a)
#   datalist[[i]] <- dat # add it to list
# }
# 
# proportion <- do.call(rbind, datalist)
# 
# orf3a$proportion <- proportion
# 
# datalist = list()
# for (i in (1:nrow(orf3a))) {
#   dat <- -sum(orf3a$proportion[i]*(log10(orf3a$proportion[i])))
#   datalist[[i]] <- dat # add it to list
# }
# 
# diversity <- do.call(rbind, datalist)
# 
# orf3a$diversity <- diversity
# 
# # combine dataframes for plotting
# combined3 <- rbind(np16, orf3a)


```

```{r}

# p1 <- ggplot(combined3, aes(x=patient, y=diversity, colour=epitope)) +
#   geom_boxplot(width=0.5) + geom_jitter(width=0.25) +
#   labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Patient", y = "Shannon diversity index")
# 
# p2 <- ggplot(combined3, aes(x=epitope, y=diversity, colour=epitope)) +
#   geom_boxplot(width=0.5)+ geom_jitter(width=0.25) +
#   labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Epitope", y = "Shannon diversity index")
# 
# p3 <- ggplot(np16, aes(x=severity, y=diversity, colour=severity)) +
#   geom_boxplot(width=0.5)+ geom_jitter(width=0.25) +
#   labs(title="Diversity of CD8+ NP16 CDR3 beta",x="Severity", y = "Shannon diversity index")


```


#### By patient
```{r, fig.width=10, fig.height=5}
# p1
```
#### By severity and epitope

```{r, fig.width=16, fig.height=7}
# grid.arrange(p2, p3, nrow=1)
```



```{r}
# # Welch 2 sample t-test: (y~x) where y is numeric and x is a binary factor
# 
# t.test(combined3$diversity~combined3$epitope)
# 
# t.test(np16$diversity~np16$severity)
# 
# # write.csv(combined2, "/t1-data/user/lfelce/TCR_analysis/diversity/cd8_np16_orf3a-28_combined_shannon.csv")

```


# NP16 and ORF3a-28 single beta cell using Isar's diversity code
Have to do table of cell name vs CDR3 beta sequence to get tabulated frequency - idea of distribution (copying BCI data example) Then can do Shannon diversity index calculation

## Diversity functions
```{r}
`diversity_function` <-
  function (x, index = "shannon", MARGIN = 1, base = exp(1))
  {
    x <- drop(as.matrix(x))
    if (!is.numeric(x))
      stop("input data must be numeric")
    if (any(x < 0, na.rm = TRUE))
      stop("input data must be non-negative")
    INDICES <- c("shannon", "simpson", "invsimpson")
    index <- match.arg(index, INDICES)
    if (length(dim(x)) > 1) {
      total <- apply(x, MARGIN, sum)
      x <- sweep(x, MARGIN, total, "/")
    } else {
      x <- x/(total <- sum(x))
    }
    if (index == "shannon")
      x <- -x * log(x, base)
    else
      x <- x*x
    if (length(dim(x)) > 1)
      H <- apply(x, MARGIN, sum, na.rm = TRUE)
    else
      H <- sum(x, na.rm = TRUE)
    if (index == "simpson")
      H <- 1 - H
    else if (index == "invsimpson")
      H <- 1/H
    ## check NA in data
    if (any(NAS <- is.na(total)))
      H[NAS] <- NA
    H
  }
```

```{r}
# test diversity functions on BCI sample data from vegan package

# library(vegan)
# data(BCI)
# 
# H <- diversity_function(BCI, "shannon")
# simp <- diversity_function(BCI, "simpson")
# invsimp <- diversity_function(BCI, "inv")
```

## Read in data, diversity calculation
Separately for NP16 and ORF3a-28 then combine for plotting
```{r, include=FALSE}
# load in data and create mixcr_b object
setwd('/t1-data/user/lfelce/TCR_analysis/cd8_np16_new/')

# create list of file names
trb_list <- list.files(path = ".", recursive = TRUE,
                        pattern = "\\TRB.txt$",
                        full.names = TRUE)
trb_list <- trb_list %>% str_replace("./*", "")

# parse in files with tcR
mixcr_b_np16 <- parse.file.list(trb_list, "mixcr")

# load in data and create  mixcr_b object
setwd('/t1-data/user/lfelce/TCR_analysis/cd8_orf_new/')

# create list of file names, remove minibulk
trb_list <- list.files(path = ".", recursive = TRUE,
                       pattern = "\\TRB.txt$",
                       full.names = TRUE)
trb_list <- trb_list %>% str_replace("./*", "")
trb_list <-trb_list[!str_detect(trb_list,pattern="minibulk")]

# parse in files with tcR
mixcr_b_orf3a <- parse.file.list(trb_list, "mixcr")

# combine mixcr files
mixcr_b <- c(mixcr_b_np16, mixcr_b_orf3a)

# sort alphabetically
mixcr_b <- mixcr_b[order(names(mixcr_b))]

# create list of cell numbers and cell names
# different lengths so same cell will be different number in a or b
mixcr_b_names <- as.data.frame(names(mixcr_b))
mixcr_b_names <- tibble::rownames_to_column(mixcr_b_names, "cell_number")

# rename columns
colnames(mixcr_b_names) <- c("cell_number", "cell_name")

```

```{r}
# mixcr_b
datalist = list()
for (i in (1:length(mixcr_b))) {
  dat <- data.frame(c(mixcr_b[[i]][3], mixcr_b[[i]][4], mixcr_b[[i]][6], mixcr_b[[i]][7], mixcr_b[[i]][8]))
  dat$i <- i # keep track of which iteration produced it
  datalist[[i]] <- dat # add it to list
}
# combine columns for each cell
# 1 beta cell only
big_data = do.call(rbind, datalist)
trb <- big_data %>% group_by(i) %>% filter(n() == 1)
colnames(trb) <- c("clone_count", "clone_fraction","CDR3_beta", "TRBV", "TRBJ", "cell_number")
trb <- merge(trb, mixcr_b_names, by="cell_number")

# tidy up
trb <- mutate(trb, beta=paste(TRBV, TRBJ, sep="_"))

# calculate frequency
table <- as.data.frame(table(trb$CDR3_beta))

# combine frequency with other info
combined2 <- merge(trb, table, by.x="CDR3_beta", by.y="Var1", all.x=TRUE)

# add metadata
combined2$patient <- ifelse(grepl("005", combined2$cell_name), "005",
                          ifelse(grepl("1131-TP-1",combined2$cell_name), "1131-TP-1",
                          ifelse(grepl("1131-TP-2", combined2$cell_name), "1131-TP-2",
                          ifelse(grepl("1153", combined2$cell_name), "1153",
                          ifelse(grepl("1201", combined2$cell_name), "1201",
                          ifelse(grepl("1105", combined2$cell_name), "1105",
                          ifelse(grepl("1134-TP-2",combined2$cell_name), "1134-TP-2",
                          ifelse(grepl("1525-TP-1", combined2$cell_name), "1525-TP-1",
                          ifelse(grepl("1525-TP-2", combined2$cell_name), "1525-TP-2",
                                 "")))))))))
combined2$severity <- ifelse(grepl("005", combined2$cell_name), "mild",
                          ifelse(grepl("1131-TP-1",combined2$cell_name), "severe",
                          ifelse(grepl("1131-TP-2", combined2$cell_name), "severe",
                          ifelse(grepl("1153", combined2$cell_name), "severe",
                          ifelse(grepl("1201", combined2$cell_name), "mild",
                          ifelse(grepl("1105", combined2$cell_name), "severe",
                          ifelse(grepl("1134-TP-2",combined2$cell_name), "severe",
                          ifelse(grepl("1525-TP-1", combined2$cell_name), "severe",
                          ifelse(grepl("1525-TP-2", combined2$cell_name), "severe",
                                 "")))))))))
combined2$timepoint <- ifelse(grepl("TP-1", combined2$patient), "acute",
                   ifelse(grepl("005|TP-2|1153|1201|1105", combined2$cell_name), "convalescent", ""))
combined2$epitope <- ifelse(grepl("NP16|B7_SPR", combined2$cell_name), "NP16",
                 ifelse(grepl("ORF3a-28", combined2$cell_name), "ORF3a-28", ""))

# subset data by epitope
np16 <- combined2[!grepl("ORF3a-28", combined2$epitope),]

orf3a <- combined2[!grepl("NP16", combined2$epitope),]

# tabulate frequencies and distribution by CDR3 sequence v cell name
table_np16 <- as.data.frame.matrix(table(np16$CDR3_beta, np16$cell_name))

table_orf3a <- as.data.frame.matrix((table(orf3a$CDR3_beta, orf3a$cell_name)))

# diversity index calculation
div_np16 <- as.data.frame(diversity_function(table_np16, "shannon"))
div_orf3a <- as.data.frame(diversity_function(table_orf3a, "shannon"))

# add back to subsetted data
div_np16 <- tibble::rownames_to_column(div_np16, "CDR3_beta")
colnames(div_np16) <- c("CDR3_beta", "diversity")

div_orf3a <- tibble::rownames_to_column(div_orf3a, "CDR3_beta")
colnames(div_orf3a) <- c("CDR3_beta", "diversity")

np16 <- merge(np16, div_np16, by="CDR3_beta", all.x=TRUE)
orf3a <- merge(orf3a, div_orf3a, by="CDR3_beta", all.x=TRUE)

# combine for plotting
combined3 <- rbind(np16, orf3a)

```

```{r}
p1 <- ggplot(combined3, aes(x=patient, y=diversity, colour=epitope)) +
  geom_boxplot(width=0.5) + geom_jitter(width=0.25) +
  labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Patient", y = "Shannon diversity index")
p2 <- ggplot(combined3, aes(x=epitope, y=diversity, colour=epitope)) +
  geom_boxplot(width=0.5)+ geom_jitter(width=0.25) +
  labs(title="Diversity of CD8+ NP16 & ORF3a-28 CDR3 beta",x="Epitope", y = "Shannon diversity index")
p3 <- ggplot(np16, aes(x=severity, y=diversity, colour=severity)) +
  geom_boxplot(width=0.5)+ geom_jitter(width=0.25) +
  labs(title="Diversity of CD8+ NP16 CDR3 beta",x="Severity", y = "Shannon diversity index")
```


## By patient
```{r, fig.width=10, fig.height=5}
p1
```
## By severity and epitope

```{r, fig.width=16, fig.height=7}
grid.arrange(p2, p3, nrow=1)
```



```{r}
# Welch 2 sample t-test: (y~x) where y is numeric and x is a binary factor
t.test(combined3$diversity~combined3$epitope)
t.test(np16$diversity~np16$severity)

write.csv(combined3, "/t1-data/user/lfelce/TCR_analysis/diversity/cd8_np16_orf3a-28_combined_shannon.csv")
```
