---
title: "TCR Repertoire Diversity"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tcR)
library(data.table)
library(stringr)
library(plyr)
library(dplyr)
library(cowplot)
library(tidyverse)
library(gridExtra)

```

TCR Repertoire Diversity looking at CDR3 sequences for alpha and beta chains separately, using MiXCR output from SmartSeq2 scRNA-Seq, and calculating diversity using Shannon diversity index (tcR package)
https://cran.r-project.org/web/packages/tcR/vignettes/tcrvignette.html

## Single cell TCRs
### CD8 NP16
```{r, include=FALSE}
# load in data and create mixcr_a and mixcr_b object
setwd('/t1-data/user/lfelce/TCR_analysis/cd8_np16_new/')

# create list of file names
tra_list <- list.files(path = ".", recursive = TRUE,
                       pattern = "\\TRA.txt$", 
                       full.names = TRUE)
tra_list <- tra_list %>% str_replace("./*", "")

trb_list <- list.files(path = ".", recursive = TRUE,
                       pattern = "\\TRB.txt$", 
                       full.names = TRUE)
trb_list <- trb_list %>% str_replace("./*", "")


# parse in files with tcR
mixcr_a <- parse.file.list(tra_list, "mixcr")
mixcr_b <- parse.file.list(trb_list, "mixcr")

# sort alphabetically
mixcr_a <- mixcr_a[order(names(mixcr_a))]
mixcr_b <- mixcr_b[order(names(mixcr_b))]

# create list of cell numbers and cell names
# different lengths so same cell will be different number in a or b
mixcr_a_names <- as.data.frame(names(mixcr_a))
mixcr_b_names <- as.data.frame(names(mixcr_b))

mixcr_a_names <-tibble::rownames_to_column(mixcr_a_names, "cell_number")
mixcr_b_names <- tibble::rownames_to_column(mixcr_b_names, "cell_number")

# rename columns
colnames(mixcr_a_names) <- c("cell_number", "cell_name")
colnames(mixcr_b_names) <- c("cell_number", "cell_name")

# convert mixcr lists to dataframe with Read count, proportion, CDR3 sequence and V and J gene info
# removed bulk samples for now
# keep cells with 1 or 2 alphas
# keep cells with 1 beta
# merge alpha and beta and keep cells that express only 1 alpha or only 1 beta

# mixcr_a
datalist = list()
for (i in (1:length(mixcr_a))) {
  dat <- data.frame(c(mixcr_a[[i]][3], mixcr_a[[i]][4], mixcr_a[[i]][6],mixcr_a[[i]][7], mixcr_a[[i]][8]))
  dat$i <- i # keep track of which iteration produced it
  datalist[[i]] <- dat # add it to list
}
# combine columns for each cell and filter to keep cells with 1 or 2 alphas
big_data = do.call(rbind, datalist)
tra <- big_data %>% group_by(i) %>% filter(n() <= 2)
colnames(tra) <- c("clone_count", "clone_fraction", "CDR3_alpha", "TRAV", "TRAJ", "cell_number")
tra <- merge(tra, mixcr_a_names, by="cell_number")


# mixcr_b
datalist = list()
for (i in (1:length(mixcr_b))) {
  dat <- data.frame(c(mixcr_b[[i]][3], mixcr_b[[i]][4], mixcr_b[[i]][6], mixcr_b[[i]][7], mixcr_b[[i]][8]))
  dat$i <- i # keep track of which iteration produced it
  datalist[[i]] <- dat # add it to list
}
# combine columns for each cell
big_data = do.call(rbind, datalist)
trb <- big_data %>% group_by(i) %>% filter(n() == 1)
colnames(trb) <- c("clone_count", "clone_fraction","CDR3_beta", "TRBV", "TRBJ", "cell_number")
trb <- merge(trb, mixcr_b_names, by="cell_number")


# combine TRA and TRB dataframes - some cells will have only 1 alpha or 1 beta, so keep all rows
cd8_np16 <- merge(tra,trb, by="cell_name", all= TRUE)

cd8_np16 <- mutate(cd8_np16, alpha=paste(TRAV, TRAJ, sep="_"))

cd8_np16 <- mutate(cd8_np16, beta=paste(TRBV, TRBJ, sep="_"))

tra <- mutate(tra, alpha=paste(TRAV, TRAJ, sep="_"))

trb <- mutate(trb, beta=paste(TRBV, TRBJ, sep="_"))
```
### Shannon diversity box plots for CDR3 alpha and beta
```{r, fig.height=10, fig.width=16, include=FALSE}
## alpha
# calculate diversity
div <- as.data.frame(repDiversity(.data = mixcr_a, .method = "entropy", .quant = "read.count"))

div_a <- tibble::rownames_to_column(div, "cell_name")
colnames(div_a ) <- c("cell_name", "diversity_index")

# add patient info and timepoint
div_a$patient <- ifelse(grepl("005", div_a$cell_name), "005", 
                          ifelse(grepl("1131-TP-1",div_a$cell_name), "1131-TP-1", 
                          ifelse(grepl("1131-TP-2", div_a$cell_name), "1131-TP-2", 
                          ifelse(grepl("1153", div_a$cell_name), "1153",
                          ifelse(grepl("1201", div_a$cell_name), "1201", "")))))

div_a$timepoint <- ifelse(grepl("1131-TP-1", div_a$patient), "acute",
                            ifelse(grepl("005|1131-TP-2|1153|1201", div_a$patient), "convalescent",""))

# merge with TCR info
div_a <- merge(div_a, tra, by="cell_name")

# add TCR status
div_a$TCR <- ifelse(grepl("TRAV21_TRAJ37|TRAV17_TRAJ22|TRAV13-1_TRAJ11|TRAV21_TRAJ30|TRAV13-1_TRAJ4", div_a$alpha), "shared",
             ifelse(grepl("TRAV13-2_TRAJ12|TRAV26-1_TRAJ4|TRAV26-1_TRAJ40", div_a$alpha), "acute only",
             ifelse(!grepl("TRAV21_TRAJ37|TRAV17_TRAJ22|TRAV13-1_TRAJ11|TRAV21_TRAJ30|TRAV13-1_TRAJ4|TRAV13-2_TRAJ12|TRAV26-1_TRAJ4|TRAV26-1_TRAJ40", div_a$alpha), "convalescent only", "")))

p1 <- ggplot(div_a, aes(x=patient, y=diversity_index, color=TCR)) + 
  geom_jitter(width=0.25) +
  labs(title="Diversity of CD8 NP16 CDR3 alpha",x="Patient", y = "Shannon diversity index") 
  

p2 <- ggplot(div_a, aes(x=timepoint, y=diversity_index, color=TCR)) + 
  geom_jitter(width=0.25)+
  labs(title="Diversity of CD8 NP16 CDR3 alpha",x="Timepoint", y = "Shannon diversity index")

# beta
div <- as.data.frame(repDiversity(.data = mixcr_b, .method = "entropy", .quant = "read.count"))

div_b <- tibble::rownames_to_column(div, "cell_name")
colnames(div_b ) <- c("cell_name", "diversity_index")

div_b$patient <- ifelse(grepl("005", div_b$cell_name), "005", 
                 ifelse(grepl("1131-TP-1",div_b$cell_name), "1131-TP-1", 
                 ifelse(grepl("1131-TP-2", div_b$cell_name), "1131-TP-2", 
                 ifelse(grepl("1153", div_b$cell_name), "1153",
                 ifelse(grepl("1201", div_b$cell_name), "1201", "")))))

div_b$timepoint <- ifelse(grepl("1131-TP-1", div_b$patient), "acute",
                   ifelse(grepl("005|1131-TP-2|1153|1201", div_b$patient), "convalescent",""))

div_b <- merge(div_b, trb, by="cell_name", all=TRUE)

p3 <- ggplot(div_b, aes(x=patient, y=diversity_index)) + 
  geom_boxplot()+
  labs(title="Diversity of CD8 NP16 CDR3 beta",x="Patient", y = "Shannon diversity index")

p4 <- ggplot(div_b, aes(x=timepoint, y=diversity_index)) + 
  geom_boxplot()+
  labs(title="Diversity of CD8 NP16 CDR3 beta",x="Timepoint", y = "Shannon diversity index")

# export files
setwd('/t1-data/user/lfelce/TCR_analysis/diversity/')
write.csv(div_a, "cd8_np16_div_a.csv")
write.csv(div_b, "cd8_np16_div_b.csv")

grid.arrange(p1, p2, p3, p4, nrow=2)

```

